{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "0279ab64",
   "metadata": {},
   "outputs": [],
   "source": [
    "#How to find, load and process snRNA-seq data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b7f1944",
   "metadata": {},
   "outputs": [],
   "source": [
    "import wget\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import anndata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e9cbd31",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Gene network analysis is a method designed to identify sub-networks (modules) of correlated genes, which are likely to be co-expressed.\n",
    "#This can be helpful in identification of sub-networks (modules) of genes that contribute to disease.\n",
    "#In this example, we will cover how to create a pairwise correlation matrix of genes, as well as how to associate them with disease."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a99b24d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#First we will cover how to find, load and process the snRNA-seq data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a56172b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Acquire snRNA-seq data from cellxgene portal: https://cellxgene.cziscience.com/collections/180bff9c-c8a5-4539-b13b-ddbc00d643e6\n",
    "#Chosen microglia cell type to focus on from this paper: Molecular characterization of selectively vulnerable neurons in Alzheimer's Disease\n",
    "#https://www.nature.com/articles/s41593-020-00764-7"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7a16480",
   "metadata": {},
   "outputs": [],
   "source": [
    "#For this tutorial, we will be using an open access freely available dataset that has been generated from microglia of the entorhinal cortex within the brain.\n",
    "#This dataset is available from the cellxgene portal, accessible here: https://cellxgene.cziscience.com/collections/180bff9c-c8a5-4539-b13b-ddbc00d643e6 entitled \"Molecular characterization of selectively vulnerable neurons in Alzheimer’s Disease: EC microglia\".\n",
    "#SnRNA-seq was performed for Controls and donors with Alzheimer's Disease.\n",
    "#This dataset was chosen due to its small size and compatability with the purpose of the pipeline.\n",
    "#This data will be available in the data/test/ directory.\n",
    "#The generated dataset is stored in h5ad format.\n",
    "#By the end of this section, we will have loaded and explored the dataset.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cf868c1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Start by downloading the dataset from the original portal.\n",
    "# URL of the dataset\n",
    "url = \"https://datasets.cellxgene.cziscience.com/1f0cd8ed-94c6-440c-bd5b-bad55e2666b1.h5ad\"\n",
    "\n",
    "# Destination path where the dataset will be saved\n",
    "destination_path = \"/shared/as8020/recode/mic_leng21.h5ad\"\n",
    "\n",
    "# Download the dataset\n",
    "wget.download(url, destination_path)\n",
    "\n",
    "#Alternatively, the dataset can be found in the dataset/test/ directory saved as mic_leng21.h5ad."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "69406b8a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Load in the test dataset\n",
    "mic = sc.read('dataset/mic_leng21.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "fa4d19c2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 5572 × 32743\n",
       "    obs: 'SampleID', 'donor_id', 'BraakStage', 'SampleBatch', 'nUMI', 'nGene', 'initialClusterAssignments', 'seurat.subclusters', 'subclusterAssignment', 'tissue_ontology_term_id', 'cell_type_ontology_term_id', 'assay_ontology_term_id', 'disease_ontology_term_id', 'self_reported_ethnicity_ontology_term_id', 'development_stage_ontology_term_id', 'sex_ontology_term_id', 'is_primary_data', 'organism_ontology_term_id', 'suspension_type', 'tissue_type', 'cell_type', 'assay', 'disease', 'organism', 'sex', 'tissue', 'self_reported_ethnicity', 'development_stage', 'observation_joinid'\n",
       "    var: 'feature_is_filtered', 'feature_name', 'feature_reference', 'feature_biotype', 'feature_length'\n",
       "    uns: 'citation', 'schema_reference', 'schema_version', 'title'\n",
       "    obsm: 'X_cca', 'X_cca.aligned', 'X_tsne'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#inspect the loaded data\n",
    "mic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "035a8f38",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature_is_filtered</th>\n",
       "      <th>feature_name</th>\n",
       "      <th>feature_reference</th>\n",
       "      <th>feature_biotype</th>\n",
       "      <th>feature_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000100568</th>\n",
       "      <td>False</td>\n",
       "      <td>VTI1B</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>gene</td>\n",
       "      <td>5768</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000101440</th>\n",
       "      <td>False</td>\n",
       "      <td>ASIP</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>gene</td>\n",
       "      <td>845</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000249847</th>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000249847.1</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>gene</td>\n",
       "      <td>646</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000136630</th>\n",
       "      <td>False</td>\n",
       "      <td>HLX</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>gene</td>\n",
       "      <td>5629</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000231731</th>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000231731.7</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>gene</td>\n",
       "      <td>5510</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000163162</th>\n",
       "      <td>False</td>\n",
       "      <td>RNF149</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>gene</td>\n",
       "      <td>5326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000185189</th>\n",
       "      <td>False</td>\n",
       "      <td>NRBP2</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>gene</td>\n",
       "      <td>4934</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000188352</th>\n",
       "      <td>False</td>\n",
       "      <td>FOCAD</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>gene</td>\n",
       "      <td>9754</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000008394</th>\n",
       "      <td>False</td>\n",
       "      <td>MGST1</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>gene</td>\n",
       "      <td>7441</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000267658</th>\n",
       "      <td>False</td>\n",
       "      <td>RAB5C-AS1</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>gene</td>\n",
       "      <td>425</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>32743 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 feature_is_filtered       feature_name feature_reference  \\\n",
       "ENSG00000100568                False              VTI1B    NCBITaxon:9606   \n",
       "ENSG00000101440                False               ASIP    NCBITaxon:9606   \n",
       "ENSG00000249847                False  ENSG00000249847.1    NCBITaxon:9606   \n",
       "ENSG00000136630                False                HLX    NCBITaxon:9606   \n",
       "ENSG00000231731                False  ENSG00000231731.7    NCBITaxon:9606   \n",
       "...                              ...                ...               ...   \n",
       "ENSG00000163162                False             RNF149    NCBITaxon:9606   \n",
       "ENSG00000185189                False              NRBP2    NCBITaxon:9606   \n",
       "ENSG00000188352                False              FOCAD    NCBITaxon:9606   \n",
       "ENSG00000008394                False              MGST1    NCBITaxon:9606   \n",
       "ENSG00000267658                False          RAB5C-AS1    NCBITaxon:9606   \n",
       "\n",
       "                feature_biotype feature_length  \n",
       "ENSG00000100568            gene           5768  \n",
       "ENSG00000101440            gene            845  \n",
       "ENSG00000249847            gene            646  \n",
       "ENSG00000136630            gene           5629  \n",
       "ENSG00000231731            gene           5510  \n",
       "...                         ...            ...  \n",
       "ENSG00000163162            gene           5326  \n",
       "ENSG00000185189            gene           4934  \n",
       "ENSG00000188352            gene           9754  \n",
       "ENSG00000008394            gene           7441  \n",
       "ENSG00000267658            gene            425  \n",
       "\n",
       "[32743 rows x 5 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Check if the gene names are in the correct format of gene symbols and not Ensembl IDs which are also common.\n",
    "mic.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "56c4913f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#As can be seen from the gene features dataframe, they have currently used the Ensembl gene naming system.\n",
    "#However, this isn't helpful for our analyses as they are not intuitively easy to interpret, instead you would need to research each Ensembl ID to identify that particular gene's name and function.\n",
    "#From the second column feature_name, it appears that the original authors have converted the Ensembl IDs to gene symbol names.\n",
    "\n",
    "#Let's go ahead and map the values in the feature_name column to the rownames of the dataframe:\n",
    "# Set the \"feature_name\" column as the index (row names)\n",
    "mic.var.set_index(\"feature_name\", drop = False, inplace=True)\n",
    "\n",
    "#It is important to note that not all Ensembl IDs map to Gene symbol names, as can be seen within rows 3 and 5 within the top of the dataframe.\n",
    "#Therefore, since there is not a mapping for all Ensembl IDs, we shall remove these rows from the dataframe as they will be difficult to interpret in subsequent analyses.\n",
    "# Filter rows where the index does not start with \"ENSG\" i.e. the Ensembl IDs.\n",
    "# Define the condition for filtering genes\n",
    "filter_genes = ~mic.var.index.str.startswith(\"ENSG\")  # Exclude genes starting with \"ENSG\"\n",
    "filter_genes\n",
    "\n",
    "# Filter genes based on the condition\n",
    "mic = mic[:, filter_genes]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "44d25469",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "View of AnnData object with n_obs × n_vars = 5572 × 24751\n",
       "    obs: 'SampleID', 'donor_id', 'BraakStage', 'SampleBatch', 'nUMI', 'nGene', 'initialClusterAssignments', 'seurat.subclusters', 'subclusterAssignment', 'tissue_ontology_term_id', 'cell_type_ontology_term_id', 'assay_ontology_term_id', 'disease_ontology_term_id', 'self_reported_ethnicity_ontology_term_id', 'development_stage_ontology_term_id', 'sex_ontology_term_id', 'is_primary_data', 'organism_ontology_term_id', 'suspension_type', 'tissue_type', 'cell_type', 'assay', 'disease', 'organism', 'sex', 'tissue', 'self_reported_ethnicity', 'development_stage', 'observation_joinid'\n",
       "    var: 'feature_is_filtered', 'feature_name', 'feature_reference', 'feature_biotype', 'feature_length'\n",
       "    uns: 'citation', 'schema_reference', 'schema_version', 'title'\n",
       "    obsm: 'X_cca', 'X_cca.aligned', 'X_tsne'"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "198693b0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature_is_filtered</th>\n",
       "      <th>feature_name</th>\n",
       "      <th>feature_reference</th>\n",
       "      <th>feature_biotype</th>\n",
       "      <th>feature_length</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>feature_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>VTI1B</th>\n",
       "      <td>False</td>\n",
       "      <td>VTI1B</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>gene</td>\n",
       "      <td>5768</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ASIP</th>\n",
       "      <td>False</td>\n",
       "      <td>ASIP</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>gene</td>\n",
       "      <td>845</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HLX</th>\n",
       "      <td>False</td>\n",
       "      <td>HLX</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>gene</td>\n",
       "      <td>5629</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C16orf87</th>\n",
       "      <td>False</td>\n",
       "      <td>C16orf87</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>gene</td>\n",
       "      <td>7216</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HMBOX1</th>\n",
       "      <td>False</td>\n",
       "      <td>HMBOX1</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>gene</td>\n",
       "      <td>6633</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNF149</th>\n",
       "      <td>False</td>\n",
       "      <td>RNF149</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>gene</td>\n",
       "      <td>5326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NRBP2</th>\n",
       "      <td>False</td>\n",
       "      <td>NRBP2</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>gene</td>\n",
       "      <td>4934</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FOCAD</th>\n",
       "      <td>False</td>\n",
       "      <td>FOCAD</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>gene</td>\n",
       "      <td>9754</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MGST1</th>\n",
       "      <td>False</td>\n",
       "      <td>MGST1</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>gene</td>\n",
       "      <td>7441</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RAB5C-AS1</th>\n",
       "      <td>False</td>\n",
       "      <td>RAB5C-AS1</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>gene</td>\n",
       "      <td>425</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>24751 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              feature_is_filtered feature_name feature_reference  \\\n",
       "feature_name                                                       \n",
       "VTI1B                       False        VTI1B    NCBITaxon:9606   \n",
       "ASIP                        False         ASIP    NCBITaxon:9606   \n",
       "HLX                         False          HLX    NCBITaxon:9606   \n",
       "C16orf87                    False     C16orf87    NCBITaxon:9606   \n",
       "HMBOX1                      False       HMBOX1    NCBITaxon:9606   \n",
       "...                           ...          ...               ...   \n",
       "RNF149                      False       RNF149    NCBITaxon:9606   \n",
       "NRBP2                       False        NRBP2    NCBITaxon:9606   \n",
       "FOCAD                       False        FOCAD    NCBITaxon:9606   \n",
       "MGST1                       False        MGST1    NCBITaxon:9606   \n",
       "RAB5C-AS1                   False    RAB5C-AS1    NCBITaxon:9606   \n",
       "\n",
       "             feature_biotype feature_length  \n",
       "feature_name                                 \n",
       "VTI1B                   gene           5768  \n",
       "ASIP                    gene            845  \n",
       "HLX                     gene           5629  \n",
       "C16orf87                gene           7216  \n",
       "HMBOX1                  gene           6633  \n",
       "...                      ...            ...  \n",
       "RNF149                  gene           5326  \n",
       "NRBP2                   gene           4934  \n",
       "FOCAD                   gene           9754  \n",
       "MGST1                   gene           7441  \n",
       "RAB5C-AS1               gene            425  \n",
       "\n",
       "[24751 rows x 5 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mic.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7615848a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#As can be seen, the number of genes have now reduced as any rows with Ensembl IDs have been removed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e469919",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Also calculate the highly variable genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56302a2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Calculating highly variable genes on gene expression data that has not been log-transformed or normalized appropriately can lead to issues, including the presence of infinity values.\n",
    "#Log transformation is a common preprocessing step for scRNA-seq data, especially when dealing with count data, to stabilize the variance and make the data more amenable to downstream analysis. \n",
    "#It helps to mitigate the impact of high expression values and reduce the influence of technical noise."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b6681a53",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/as8020/.local/lib/python3.8/site-packages/scanpy/preprocessing/_simple.py:284: DeprecationWarning: Use is_view instead of isview, isview will be removed in the future.\n",
      "  if isinstance(data, AnnData) and data.isview:\n"
     ]
    }
   ],
   "source": [
    "# Log normalize the gene expression data\n",
    "sc.pp.log1p(mic)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "20e24dfd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate highly variable genes\n",
    "sc.pp.highly_variable_genes(mic, n_top_genes = 1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "25b9a63a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 5572 × 24751\n",
       "    obs: 'SampleID', 'donor_id', 'BraakStage', 'SampleBatch', 'nUMI', 'nGene', 'initialClusterAssignments', 'seurat.subclusters', 'subclusterAssignment', 'tissue_ontology_term_id', 'cell_type_ontology_term_id', 'assay_ontology_term_id', 'disease_ontology_term_id', 'self_reported_ethnicity_ontology_term_id', 'development_stage_ontology_term_id', 'sex_ontology_term_id', 'is_primary_data', 'organism_ontology_term_id', 'suspension_type', 'tissue_type', 'cell_type', 'assay', 'disease', 'organism', 'sex', 'tissue', 'self_reported_ethnicity', 'development_stage', 'observation_joinid'\n",
       "    var: 'feature_is_filtered', 'feature_name', 'feature_reference', 'feature_biotype', 'feature_length', 'highly_variable', 'means', 'dispersions', 'dispersions_norm'\n",
       "    uns: 'citation', 'schema_reference', 'schema_version', 'title'\n",
       "    obsm: 'X_cca', 'X_cca.aligned', 'X_tsne'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "142cabfe",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Lets save the filtered object\n",
    "mic.write_h5ad('dataset/mic_leng21_filtered.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aef1ce9d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "973ca838",
   "metadata": {},
   "outputs": [],
   "source": [
    "#We will now explore the associated metadata "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "cb920eda",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['SampleID', 'donor_id', 'BraakStage', 'SampleBatch', 'nUMI', 'nGene',\n",
       "       'initialClusterAssignments', 'seurat.subclusters',\n",
       "       'subclusterAssignment', 'tissue_ontology_term_id',\n",
       "       'cell_type_ontology_term_id', 'assay_ontology_term_id',\n",
       "       'disease_ontology_term_id', 'self_reported_ethnicity_ontology_term_id',\n",
       "       'development_stage_ontology_term_id', 'sex_ontology_term_id',\n",
       "       'is_primary_data', 'organism_ontology_term_id', 'suspension_type',\n",
       "       'tissue_type', 'cell_type', 'assay', 'disease', 'organism', 'sex',\n",
       "       'tissue', 'self_reported_ethnicity', 'development_stage',\n",
       "       'observation_joinid'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mic.obs.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "543e94cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "#As can be seen, this dataset contains 5572 cells and 32743 genes.\n",
    "#It also has relevant metadata in the obs section, such as BraakStage. \n",
    "#The metadata may need to be encoded into the correct format for subsequent analyses, so let's have a look at the current format."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "95d95d35",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SampleID</th>\n",
       "      <th>donor_id</th>\n",
       "      <th>BraakStage</th>\n",
       "      <th>SampleBatch</th>\n",
       "      <th>nUMI</th>\n",
       "      <th>nGene</th>\n",
       "      <th>initialClusterAssignments</th>\n",
       "      <th>seurat.subclusters</th>\n",
       "      <th>subclusterAssignment</th>\n",
       "      <th>tissue_ontology_term_id</th>\n",
       "      <th>...</th>\n",
       "      <th>tissue_type</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>assay</th>\n",
       "      <th>disease</th>\n",
       "      <th>organism</th>\n",
       "      <th>sex</th>\n",
       "      <th>tissue</th>\n",
       "      <th>self_reported_ethnicity</th>\n",
       "      <th>development_stage</th>\n",
       "      <th>observation_joinid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>EC3_AACCATGTCACGAAGG</th>\n",
       "      <td>EC3</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>C</td>\n",
       "      <td>219.0</td>\n",
       "      <td>193</td>\n",
       "      <td>EC:c1</td>\n",
       "      <td>0</td>\n",
       "      <td>EC:Micro.s0</td>\n",
       "      <td>UBERON:0002728</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue</td>\n",
       "      <td>mature microglial cell</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>normal</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>male</td>\n",
       "      <td>entorhinal cortex</td>\n",
       "      <td>unknown</td>\n",
       "      <td>71-year-old human stage</td>\n",
       "      <td>K#V~xdAtWl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC3_AATCGGTCACTTACGA</th>\n",
       "      <td>EC3</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>C</td>\n",
       "      <td>266.0</td>\n",
       "      <td>223</td>\n",
       "      <td>EC:c6</td>\n",
       "      <td>1</td>\n",
       "      <td>EC:Micro.s1</td>\n",
       "      <td>UBERON:0002728</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue</td>\n",
       "      <td>mature microglial cell</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>normal</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>male</td>\n",
       "      <td>entorhinal cortex</td>\n",
       "      <td>unknown</td>\n",
       "      <td>71-year-old human stage</td>\n",
       "      <td>SF!nEev15F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC3_ACAGCTAAGTGTCCCG</th>\n",
       "      <td>EC3</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>C</td>\n",
       "      <td>233.0</td>\n",
       "      <td>204</td>\n",
       "      <td>EC:c1</td>\n",
       "      <td>1</td>\n",
       "      <td>EC:Micro.s1</td>\n",
       "      <td>UBERON:0002728</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue</td>\n",
       "      <td>mature microglial cell</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>normal</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>male</td>\n",
       "      <td>entorhinal cortex</td>\n",
       "      <td>unknown</td>\n",
       "      <td>71-year-old human stage</td>\n",
       "      <td>-EP32!nM28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC3_ACAGCTACAAGGACAC</th>\n",
       "      <td>EC3</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>C</td>\n",
       "      <td>285.0</td>\n",
       "      <td>242</td>\n",
       "      <td>EC:c1</td>\n",
       "      <td>1</td>\n",
       "      <td>EC:Micro.s1</td>\n",
       "      <td>UBERON:0002728</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue</td>\n",
       "      <td>mature microglial cell</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>normal</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>male</td>\n",
       "      <td>entorhinal cortex</td>\n",
       "      <td>unknown</td>\n",
       "      <td>71-year-old human stage</td>\n",
       "      <td>LW&lt;Bj-F)!}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC3_ACAGCTAGTGATGTCT</th>\n",
       "      <td>EC3</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>C</td>\n",
       "      <td>959.0</td>\n",
       "      <td>699</td>\n",
       "      <td>EC:c5</td>\n",
       "      <td>2</td>\n",
       "      <td>EC:Micro.s2</td>\n",
       "      <td>UBERON:0002728</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue</td>\n",
       "      <td>mature microglial cell</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>normal</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>male</td>\n",
       "      <td>entorhinal cortex</td>\n",
       "      <td>unknown</td>\n",
       "      <td>71-year-old human stage</td>\n",
       "      <td>xqpir=h}cz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC10_TTTCCTCGTGGTAACG</th>\n",
       "      <td>EC10</td>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>D</td>\n",
       "      <td>1010.0</td>\n",
       "      <td>710</td>\n",
       "      <td>EC:c1</td>\n",
       "      <td>1</td>\n",
       "      <td>EC:Micro.s1</td>\n",
       "      <td>UBERON:0002728</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue</td>\n",
       "      <td>mature microglial cell</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>Alzheimer disease</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>male</td>\n",
       "      <td>entorhinal cortex</td>\n",
       "      <td>unknown</td>\n",
       "      <td>82-year-old human stage</td>\n",
       "      <td>YGUBOx%{wc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC10_TTTCCTCGTTTGACAC</th>\n",
       "      <td>EC10</td>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>D</td>\n",
       "      <td>1355.0</td>\n",
       "      <td>1013</td>\n",
       "      <td>EC:c1</td>\n",
       "      <td>1</td>\n",
       "      <td>EC:Micro.s1</td>\n",
       "      <td>UBERON:0002728</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue</td>\n",
       "      <td>mature microglial cell</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>Alzheimer disease</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>male</td>\n",
       "      <td>entorhinal cortex</td>\n",
       "      <td>unknown</td>\n",
       "      <td>82-year-old human stage</td>\n",
       "      <td>1Ns|7j@C8q</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC10_TTTCCTCTCACAGGCC</th>\n",
       "      <td>EC10</td>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>D</td>\n",
       "      <td>1971.0</td>\n",
       "      <td>1358</td>\n",
       "      <td>EC:c1</td>\n",
       "      <td>0</td>\n",
       "      <td>EC:Micro.s0</td>\n",
       "      <td>UBERON:0002728</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue</td>\n",
       "      <td>mature microglial cell</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>Alzheimer disease</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>male</td>\n",
       "      <td>entorhinal cortex</td>\n",
       "      <td>unknown</td>\n",
       "      <td>82-year-old human stage</td>\n",
       "      <td>=JuQ#I!6*4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC10_TTTGTCACAAGCGTAG</th>\n",
       "      <td>EC10</td>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>D</td>\n",
       "      <td>403.0</td>\n",
       "      <td>342</td>\n",
       "      <td>EC:c1</td>\n",
       "      <td>2</td>\n",
       "      <td>EC:Micro.s2</td>\n",
       "      <td>UBERON:0002728</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue</td>\n",
       "      <td>mature microglial cell</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>Alzheimer disease</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>male</td>\n",
       "      <td>entorhinal cortex</td>\n",
       "      <td>unknown</td>\n",
       "      <td>82-year-old human stage</td>\n",
       "      <td>GA7~u!F(fS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC10_TTTGTCATCTATCGCC</th>\n",
       "      <td>EC10</td>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>D</td>\n",
       "      <td>444.0</td>\n",
       "      <td>393</td>\n",
       "      <td>EC:c1</td>\n",
       "      <td>2</td>\n",
       "      <td>EC:Micro.s2</td>\n",
       "      <td>UBERON:0002728</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue</td>\n",
       "      <td>mature microglial cell</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>Alzheimer disease</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>male</td>\n",
       "      <td>entorhinal cortex</td>\n",
       "      <td>unknown</td>\n",
       "      <td>82-year-old human stage</td>\n",
       "      <td>Zpfx-&gt;svc&gt;</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5572 rows × 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      SampleID donor_id BraakStage SampleBatch    nUMI  nGene  \\\n",
       "EC3_AACCATGTCACGAAGG       EC3        3          0           C   219.0    193   \n",
       "EC3_AATCGGTCACTTACGA       EC3        3          0           C   266.0    223   \n",
       "EC3_ACAGCTAAGTGTCCCG       EC3        3          0           C   233.0    204   \n",
       "EC3_ACAGCTACAAGGACAC       EC3        3          0           C   285.0    242   \n",
       "EC3_ACAGCTAGTGATGTCT       EC3        3          0           C   959.0    699   \n",
       "...                        ...      ...        ...         ...     ...    ...   \n",
       "EC10_TTTCCTCGTGGTAACG     EC10       10          6           D  1010.0    710   \n",
       "EC10_TTTCCTCGTTTGACAC     EC10       10          6           D  1355.0   1013   \n",
       "EC10_TTTCCTCTCACAGGCC     EC10       10          6           D  1971.0   1358   \n",
       "EC10_TTTGTCACAAGCGTAG     EC10       10          6           D   403.0    342   \n",
       "EC10_TTTGTCATCTATCGCC     EC10       10          6           D   444.0    393   \n",
       "\n",
       "                      initialClusterAssignments seurat.subclusters  \\\n",
       "EC3_AACCATGTCACGAAGG                      EC:c1                  0   \n",
       "EC3_AATCGGTCACTTACGA                      EC:c6                  1   \n",
       "EC3_ACAGCTAAGTGTCCCG                      EC:c1                  1   \n",
       "EC3_ACAGCTACAAGGACAC                      EC:c1                  1   \n",
       "EC3_ACAGCTAGTGATGTCT                      EC:c5                  2   \n",
       "...                                         ...                ...   \n",
       "EC10_TTTCCTCGTGGTAACG                     EC:c1                  1   \n",
       "EC10_TTTCCTCGTTTGACAC                     EC:c1                  1   \n",
       "EC10_TTTCCTCTCACAGGCC                     EC:c1                  0   \n",
       "EC10_TTTGTCACAAGCGTAG                     EC:c1                  2   \n",
       "EC10_TTTGTCATCTATCGCC                     EC:c1                  2   \n",
       "\n",
       "                      subclusterAssignment tissue_ontology_term_id  ...  \\\n",
       "EC3_AACCATGTCACGAAGG           EC:Micro.s0          UBERON:0002728  ...   \n",
       "EC3_AATCGGTCACTTACGA           EC:Micro.s1          UBERON:0002728  ...   \n",
       "EC3_ACAGCTAAGTGTCCCG           EC:Micro.s1          UBERON:0002728  ...   \n",
       "EC3_ACAGCTACAAGGACAC           EC:Micro.s1          UBERON:0002728  ...   \n",
       "EC3_ACAGCTAGTGATGTCT           EC:Micro.s2          UBERON:0002728  ...   \n",
       "...                                    ...                     ...  ...   \n",
       "EC10_TTTCCTCGTGGTAACG          EC:Micro.s1          UBERON:0002728  ...   \n",
       "EC10_TTTCCTCGTTTGACAC          EC:Micro.s1          UBERON:0002728  ...   \n",
       "EC10_TTTCCTCTCACAGGCC          EC:Micro.s0          UBERON:0002728  ...   \n",
       "EC10_TTTGTCACAAGCGTAG          EC:Micro.s2          UBERON:0002728  ...   \n",
       "EC10_TTTGTCATCTATCGCC          EC:Micro.s2          UBERON:0002728  ...   \n",
       "\n",
       "                      tissue_type               cell_type      assay  \\\n",
       "EC3_AACCATGTCACGAAGG       tissue  mature microglial cell  10x 3' v2   \n",
       "EC3_AATCGGTCACTTACGA       tissue  mature microglial cell  10x 3' v2   \n",
       "EC3_ACAGCTAAGTGTCCCG       tissue  mature microglial cell  10x 3' v2   \n",
       "EC3_ACAGCTACAAGGACAC       tissue  mature microglial cell  10x 3' v2   \n",
       "EC3_ACAGCTAGTGATGTCT       tissue  mature microglial cell  10x 3' v2   \n",
       "...                           ...                     ...        ...   \n",
       "EC10_TTTCCTCGTGGTAACG      tissue  mature microglial cell  10x 3' v2   \n",
       "EC10_TTTCCTCGTTTGACAC      tissue  mature microglial cell  10x 3' v2   \n",
       "EC10_TTTCCTCTCACAGGCC      tissue  mature microglial cell  10x 3' v2   \n",
       "EC10_TTTGTCACAAGCGTAG      tissue  mature microglial cell  10x 3' v2   \n",
       "EC10_TTTGTCATCTATCGCC      tissue  mature microglial cell  10x 3' v2   \n",
       "\n",
       "                                 disease      organism   sex  \\\n",
       "EC3_AACCATGTCACGAAGG              normal  Homo sapiens  male   \n",
       "EC3_AATCGGTCACTTACGA              normal  Homo sapiens  male   \n",
       "EC3_ACAGCTAAGTGTCCCG              normal  Homo sapiens  male   \n",
       "EC3_ACAGCTACAAGGACAC              normal  Homo sapiens  male   \n",
       "EC3_ACAGCTAGTGATGTCT              normal  Homo sapiens  male   \n",
       "...                                  ...           ...   ...   \n",
       "EC10_TTTCCTCGTGGTAACG  Alzheimer disease  Homo sapiens  male   \n",
       "EC10_TTTCCTCGTTTGACAC  Alzheimer disease  Homo sapiens  male   \n",
       "EC10_TTTCCTCTCACAGGCC  Alzheimer disease  Homo sapiens  male   \n",
       "EC10_TTTGTCACAAGCGTAG  Alzheimer disease  Homo sapiens  male   \n",
       "EC10_TTTGTCATCTATCGCC  Alzheimer disease  Homo sapiens  male   \n",
       "\n",
       "                                  tissue self_reported_ethnicity  \\\n",
       "EC3_AACCATGTCACGAAGG   entorhinal cortex                 unknown   \n",
       "EC3_AATCGGTCACTTACGA   entorhinal cortex                 unknown   \n",
       "EC3_ACAGCTAAGTGTCCCG   entorhinal cortex                 unknown   \n",
       "EC3_ACAGCTACAAGGACAC   entorhinal cortex                 unknown   \n",
       "EC3_ACAGCTAGTGATGTCT   entorhinal cortex                 unknown   \n",
       "...                                  ...                     ...   \n",
       "EC10_TTTCCTCGTGGTAACG  entorhinal cortex                 unknown   \n",
       "EC10_TTTCCTCGTTTGACAC  entorhinal cortex                 unknown   \n",
       "EC10_TTTCCTCTCACAGGCC  entorhinal cortex                 unknown   \n",
       "EC10_TTTGTCACAAGCGTAG  entorhinal cortex                 unknown   \n",
       "EC10_TTTGTCATCTATCGCC  entorhinal cortex                 unknown   \n",
       "\n",
       "                             development_stage observation_joinid  \n",
       "EC3_AACCATGTCACGAAGG   71-year-old human stage         K#V~xdAtWl  \n",
       "EC3_AATCGGTCACTTACGA   71-year-old human stage         SF!nEev15F  \n",
       "EC3_ACAGCTAAGTGTCCCG   71-year-old human stage         -EP32!nM28  \n",
       "EC3_ACAGCTACAAGGACAC   71-year-old human stage         LW<Bj-F)!}  \n",
       "EC3_ACAGCTAGTGATGTCT   71-year-old human stage         xqpir=h}cz  \n",
       "...                                        ...                ...  \n",
       "EC10_TTTCCTCGTGGTAACG  82-year-old human stage         YGUBOx%{wc  \n",
       "EC10_TTTCCTCGTTTGACAC  82-year-old human stage         1Ns|7j@C8q  \n",
       "EC10_TTTCCTCTCACAGGCC  82-year-old human stage         =JuQ#I!6*4  \n",
       "EC10_TTTGTCACAAGCGTAG  82-year-old human stage         GA7~u!F(fS  \n",
       "EC10_TTTGTCATCTATCGCC  82-year-old human stage         Zpfx->svc>  \n",
       "\n",
       "[5572 rows x 29 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mic.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "b67ee215",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SampleID</th>\n",
       "      <th>donor_id</th>\n",
       "      <th>BraakStage</th>\n",
       "      <th>SampleBatch</th>\n",
       "      <th>nUMI</th>\n",
       "      <th>nGene</th>\n",
       "      <th>initialClusterAssignments</th>\n",
       "      <th>seurat.subclusters</th>\n",
       "      <th>subclusterAssignment</th>\n",
       "      <th>tissue_ontology_term_id</th>\n",
       "      <th>...</th>\n",
       "      <th>tissue_type</th>\n",
       "      <th>cell_type</th>\n",
       "      <th>assay</th>\n",
       "      <th>disease</th>\n",
       "      <th>organism</th>\n",
       "      <th>sex</th>\n",
       "      <th>tissue</th>\n",
       "      <th>self_reported_ethnicity</th>\n",
       "      <th>development_stage</th>\n",
       "      <th>observation_joinid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>EC3_AACCATGTCACGAAGG</th>\n",
       "      <td>EC3</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>C</td>\n",
       "      <td>219.0</td>\n",
       "      <td>193</td>\n",
       "      <td>EC:c1</td>\n",
       "      <td>0</td>\n",
       "      <td>EC:Micro.s0</td>\n",
       "      <td>UBERON:0002728</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue</td>\n",
       "      <td>mature microglial cell</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>normal</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>male</td>\n",
       "      <td>entorhinal cortex</td>\n",
       "      <td>unknown</td>\n",
       "      <td>71-year-old human stage</td>\n",
       "      <td>K#V~xdAtWl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC3_AATCGGTCACTTACGA</th>\n",
       "      <td>EC3</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>C</td>\n",
       "      <td>266.0</td>\n",
       "      <td>223</td>\n",
       "      <td>EC:c6</td>\n",
       "      <td>1</td>\n",
       "      <td>EC:Micro.s1</td>\n",
       "      <td>UBERON:0002728</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue</td>\n",
       "      <td>mature microglial cell</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>normal</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>male</td>\n",
       "      <td>entorhinal cortex</td>\n",
       "      <td>unknown</td>\n",
       "      <td>71-year-old human stage</td>\n",
       "      <td>SF!nEev15F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC3_ACAGCTAAGTGTCCCG</th>\n",
       "      <td>EC3</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>C</td>\n",
       "      <td>233.0</td>\n",
       "      <td>204</td>\n",
       "      <td>EC:c1</td>\n",
       "      <td>1</td>\n",
       "      <td>EC:Micro.s1</td>\n",
       "      <td>UBERON:0002728</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue</td>\n",
       "      <td>mature microglial cell</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>normal</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>male</td>\n",
       "      <td>entorhinal cortex</td>\n",
       "      <td>unknown</td>\n",
       "      <td>71-year-old human stage</td>\n",
       "      <td>-EP32!nM28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC3_ACAGCTACAAGGACAC</th>\n",
       "      <td>EC3</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>C</td>\n",
       "      <td>285.0</td>\n",
       "      <td>242</td>\n",
       "      <td>EC:c1</td>\n",
       "      <td>1</td>\n",
       "      <td>EC:Micro.s1</td>\n",
       "      <td>UBERON:0002728</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue</td>\n",
       "      <td>mature microglial cell</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>normal</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>male</td>\n",
       "      <td>entorhinal cortex</td>\n",
       "      <td>unknown</td>\n",
       "      <td>71-year-old human stage</td>\n",
       "      <td>LW&lt;Bj-F)!}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC3_ACAGCTAGTGATGTCT</th>\n",
       "      <td>EC3</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>C</td>\n",
       "      <td>959.0</td>\n",
       "      <td>699</td>\n",
       "      <td>EC:c5</td>\n",
       "      <td>2</td>\n",
       "      <td>EC:Micro.s2</td>\n",
       "      <td>UBERON:0002728</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue</td>\n",
       "      <td>mature microglial cell</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>normal</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>male</td>\n",
       "      <td>entorhinal cortex</td>\n",
       "      <td>unknown</td>\n",
       "      <td>71-year-old human stage</td>\n",
       "      <td>xqpir=h}cz</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC10_TTTCCTCGTGGTAACG</th>\n",
       "      <td>EC10</td>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>D</td>\n",
       "      <td>1010.0</td>\n",
       "      <td>710</td>\n",
       "      <td>EC:c1</td>\n",
       "      <td>1</td>\n",
       "      <td>EC:Micro.s1</td>\n",
       "      <td>UBERON:0002728</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue</td>\n",
       "      <td>mature microglial cell</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>Alzheimer disease</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>male</td>\n",
       "      <td>entorhinal cortex</td>\n",
       "      <td>unknown</td>\n",
       "      <td>82-year-old human stage</td>\n",
       "      <td>YGUBOx%{wc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC10_TTTCCTCGTTTGACAC</th>\n",
       "      <td>EC10</td>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>D</td>\n",
       "      <td>1355.0</td>\n",
       "      <td>1013</td>\n",
       "      <td>EC:c1</td>\n",
       "      <td>1</td>\n",
       "      <td>EC:Micro.s1</td>\n",
       "      <td>UBERON:0002728</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue</td>\n",
       "      <td>mature microglial cell</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>Alzheimer disease</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>male</td>\n",
       "      <td>entorhinal cortex</td>\n",
       "      <td>unknown</td>\n",
       "      <td>82-year-old human stage</td>\n",
       "      <td>1Ns|7j@C8q</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC10_TTTCCTCTCACAGGCC</th>\n",
       "      <td>EC10</td>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>D</td>\n",
       "      <td>1971.0</td>\n",
       "      <td>1358</td>\n",
       "      <td>EC:c1</td>\n",
       "      <td>0</td>\n",
       "      <td>EC:Micro.s0</td>\n",
       "      <td>UBERON:0002728</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue</td>\n",
       "      <td>mature microglial cell</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>Alzheimer disease</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>male</td>\n",
       "      <td>entorhinal cortex</td>\n",
       "      <td>unknown</td>\n",
       "      <td>82-year-old human stage</td>\n",
       "      <td>=JuQ#I!6*4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC10_TTTGTCACAAGCGTAG</th>\n",
       "      <td>EC10</td>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>D</td>\n",
       "      <td>403.0</td>\n",
       "      <td>342</td>\n",
       "      <td>EC:c1</td>\n",
       "      <td>2</td>\n",
       "      <td>EC:Micro.s2</td>\n",
       "      <td>UBERON:0002728</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue</td>\n",
       "      <td>mature microglial cell</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>Alzheimer disease</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>male</td>\n",
       "      <td>entorhinal cortex</td>\n",
       "      <td>unknown</td>\n",
       "      <td>82-year-old human stage</td>\n",
       "      <td>GA7~u!F(fS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC10_TTTGTCATCTATCGCC</th>\n",
       "      <td>EC10</td>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>D</td>\n",
       "      <td>444.0</td>\n",
       "      <td>393</td>\n",
       "      <td>EC:c1</td>\n",
       "      <td>2</td>\n",
       "      <td>EC:Micro.s2</td>\n",
       "      <td>UBERON:0002728</td>\n",
       "      <td>...</td>\n",
       "      <td>tissue</td>\n",
       "      <td>mature microglial cell</td>\n",
       "      <td>10x 3' v2</td>\n",
       "      <td>Alzheimer disease</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>male</td>\n",
       "      <td>entorhinal cortex</td>\n",
       "      <td>unknown</td>\n",
       "      <td>82-year-old human stage</td>\n",
       "      <td>Zpfx-&gt;svc&gt;</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5572 rows × 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      SampleID donor_id BraakStage SampleBatch    nUMI  nGene  \\\n",
       "EC3_AACCATGTCACGAAGG       EC3        3          0           C   219.0    193   \n",
       "EC3_AATCGGTCACTTACGA       EC3        3          0           C   266.0    223   \n",
       "EC3_ACAGCTAAGTGTCCCG       EC3        3          0           C   233.0    204   \n",
       "EC3_ACAGCTACAAGGACAC       EC3        3          0           C   285.0    242   \n",
       "EC3_ACAGCTAGTGATGTCT       EC3        3          0           C   959.0    699   \n",
       "...                        ...      ...        ...         ...     ...    ...   \n",
       "EC10_TTTCCTCGTGGTAACG     EC10       10          6           D  1010.0    710   \n",
       "EC10_TTTCCTCGTTTGACAC     EC10       10          6           D  1355.0   1013   \n",
       "EC10_TTTCCTCTCACAGGCC     EC10       10          6           D  1971.0   1358   \n",
       "EC10_TTTGTCACAAGCGTAG     EC10       10          6           D   403.0    342   \n",
       "EC10_TTTGTCATCTATCGCC     EC10       10          6           D   444.0    393   \n",
       "\n",
       "                      initialClusterAssignments seurat.subclusters  \\\n",
       "EC3_AACCATGTCACGAAGG                      EC:c1                  0   \n",
       "EC3_AATCGGTCACTTACGA                      EC:c6                  1   \n",
       "EC3_ACAGCTAAGTGTCCCG                      EC:c1                  1   \n",
       "EC3_ACAGCTACAAGGACAC                      EC:c1                  1   \n",
       "EC3_ACAGCTAGTGATGTCT                      EC:c5                  2   \n",
       "...                                         ...                ...   \n",
       "EC10_TTTCCTCGTGGTAACG                     EC:c1                  1   \n",
       "EC10_TTTCCTCGTTTGACAC                     EC:c1                  1   \n",
       "EC10_TTTCCTCTCACAGGCC                     EC:c1                  0   \n",
       "EC10_TTTGTCACAAGCGTAG                     EC:c1                  2   \n",
       "EC10_TTTGTCATCTATCGCC                     EC:c1                  2   \n",
       "\n",
       "                      subclusterAssignment tissue_ontology_term_id  ...  \\\n",
       "EC3_AACCATGTCACGAAGG           EC:Micro.s0          UBERON:0002728  ...   \n",
       "EC3_AATCGGTCACTTACGA           EC:Micro.s1          UBERON:0002728  ...   \n",
       "EC3_ACAGCTAAGTGTCCCG           EC:Micro.s1          UBERON:0002728  ...   \n",
       "EC3_ACAGCTACAAGGACAC           EC:Micro.s1          UBERON:0002728  ...   \n",
       "EC3_ACAGCTAGTGATGTCT           EC:Micro.s2          UBERON:0002728  ...   \n",
       "...                                    ...                     ...  ...   \n",
       "EC10_TTTCCTCGTGGTAACG          EC:Micro.s1          UBERON:0002728  ...   \n",
       "EC10_TTTCCTCGTTTGACAC          EC:Micro.s1          UBERON:0002728  ...   \n",
       "EC10_TTTCCTCTCACAGGCC          EC:Micro.s0          UBERON:0002728  ...   \n",
       "EC10_TTTGTCACAAGCGTAG          EC:Micro.s2          UBERON:0002728  ...   \n",
       "EC10_TTTGTCATCTATCGCC          EC:Micro.s2          UBERON:0002728  ...   \n",
       "\n",
       "                      tissue_type               cell_type      assay  \\\n",
       "EC3_AACCATGTCACGAAGG       tissue  mature microglial cell  10x 3' v2   \n",
       "EC3_AATCGGTCACTTACGA       tissue  mature microglial cell  10x 3' v2   \n",
       "EC3_ACAGCTAAGTGTCCCG       tissue  mature microglial cell  10x 3' v2   \n",
       "EC3_ACAGCTACAAGGACAC       tissue  mature microglial cell  10x 3' v2   \n",
       "EC3_ACAGCTAGTGATGTCT       tissue  mature microglial cell  10x 3' v2   \n",
       "...                           ...                     ...        ...   \n",
       "EC10_TTTCCTCGTGGTAACG      tissue  mature microglial cell  10x 3' v2   \n",
       "EC10_TTTCCTCGTTTGACAC      tissue  mature microglial cell  10x 3' v2   \n",
       "EC10_TTTCCTCTCACAGGCC      tissue  mature microglial cell  10x 3' v2   \n",
       "EC10_TTTGTCACAAGCGTAG      tissue  mature microglial cell  10x 3' v2   \n",
       "EC10_TTTGTCATCTATCGCC      tissue  mature microglial cell  10x 3' v2   \n",
       "\n",
       "                                 disease      organism   sex  \\\n",
       "EC3_AACCATGTCACGAAGG              normal  Homo sapiens  male   \n",
       "EC3_AATCGGTCACTTACGA              normal  Homo sapiens  male   \n",
       "EC3_ACAGCTAAGTGTCCCG              normal  Homo sapiens  male   \n",
       "EC3_ACAGCTACAAGGACAC              normal  Homo sapiens  male   \n",
       "EC3_ACAGCTAGTGATGTCT              normal  Homo sapiens  male   \n",
       "...                                  ...           ...   ...   \n",
       "EC10_TTTCCTCGTGGTAACG  Alzheimer disease  Homo sapiens  male   \n",
       "EC10_TTTCCTCGTTTGACAC  Alzheimer disease  Homo sapiens  male   \n",
       "EC10_TTTCCTCTCACAGGCC  Alzheimer disease  Homo sapiens  male   \n",
       "EC10_TTTGTCACAAGCGTAG  Alzheimer disease  Homo sapiens  male   \n",
       "EC10_TTTGTCATCTATCGCC  Alzheimer disease  Homo sapiens  male   \n",
       "\n",
       "                                  tissue self_reported_ethnicity  \\\n",
       "EC3_AACCATGTCACGAAGG   entorhinal cortex                 unknown   \n",
       "EC3_AATCGGTCACTTACGA   entorhinal cortex                 unknown   \n",
       "EC3_ACAGCTAAGTGTCCCG   entorhinal cortex                 unknown   \n",
       "EC3_ACAGCTACAAGGACAC   entorhinal cortex                 unknown   \n",
       "EC3_ACAGCTAGTGATGTCT   entorhinal cortex                 unknown   \n",
       "...                                  ...                     ...   \n",
       "EC10_TTTCCTCGTGGTAACG  entorhinal cortex                 unknown   \n",
       "EC10_TTTCCTCGTTTGACAC  entorhinal cortex                 unknown   \n",
       "EC10_TTTCCTCTCACAGGCC  entorhinal cortex                 unknown   \n",
       "EC10_TTTGTCACAAGCGTAG  entorhinal cortex                 unknown   \n",
       "EC10_TTTGTCATCTATCGCC  entorhinal cortex                 unknown   \n",
       "\n",
       "                             development_stage observation_joinid  \n",
       "EC3_AACCATGTCACGAAGG   71-year-old human stage         K#V~xdAtWl  \n",
       "EC3_AATCGGTCACTTACGA   71-year-old human stage         SF!nEev15F  \n",
       "EC3_ACAGCTAAGTGTCCCG   71-year-old human stage         -EP32!nM28  \n",
       "EC3_ACAGCTACAAGGACAC   71-year-old human stage         LW<Bj-F)!}  \n",
       "EC3_ACAGCTAGTGATGTCT   71-year-old human stage         xqpir=h}cz  \n",
       "...                                        ...                ...  \n",
       "EC10_TTTCCTCGTGGTAACG  82-year-old human stage         YGUBOx%{wc  \n",
       "EC10_TTTCCTCGTTTGACAC  82-year-old human stage         1Ns|7j@C8q  \n",
       "EC10_TTTCCTCTCACAGGCC  82-year-old human stage         =JuQ#I!6*4  \n",
       "EC10_TTTGTCACAAGCGTAG  82-year-old human stage         GA7~u!F(fS  \n",
       "EC10_TTTGTCATCTATCGCC  82-year-old human stage         Zpfx->svc>  \n",
       "\n",
       "[5572 rows x 29 columns]"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Lets create a separate dataframe with the metadata information as this will be needed for the correlation analysis.\n",
    "#Currently we want to create a copy of the metadata so as not to alter the original adata object.\n",
    "metadata = mic.obs.copy()\n",
    "metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "c6cb77f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#There are many columns that are not needed.\n",
    "#Let's remove uninteresting columns\n",
    "columns_to_remove = ['SampleID', 'SampleBatch',\n",
    "       'initialClusterAssignments',\n",
    "       'subclusterAssignment', 'tissue_ontology_term_id',\n",
    "       'cell_type_ontology_term_id', 'assay_ontology_term_id',\n",
    "       'disease_ontology_term_id', 'self_reported_ethnicity_ontology_term_id',\n",
    "       'development_stage_ontology_term_id', 'sex_ontology_term_id',\n",
    "       'is_primary_data', 'organism_ontology_term_id', 'suspension_type',\n",
    "       'tissue_type', 'cell_type', 'assay', 'organism',\n",
    "       'tissue', 'self_reported_ethnicity',\n",
    "       'observation_joinid' ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "e1e57a68",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata.drop(columns=columns_to_remove, inplace = True) #Set inplace=True to modify the DataFrame in place. If you set inplace=False or omit it, the drop() method will return a new DataFrame with the specified columns removed, leaving the original DataFrame unchanged."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "472b3a05",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>donor_id</th>\n",
       "      <th>BraakStage</th>\n",
       "      <th>nUMI</th>\n",
       "      <th>nGene</th>\n",
       "      <th>seurat.subclusters</th>\n",
       "      <th>disease</th>\n",
       "      <th>sex</th>\n",
       "      <th>development_stage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>EC3_AACCATGTCACGAAGG</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>219.0</td>\n",
       "      <td>193</td>\n",
       "      <td>0</td>\n",
       "      <td>normal</td>\n",
       "      <td>male</td>\n",
       "      <td>71-year-old human stage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC3_AATCGGTCACTTACGA</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>266.0</td>\n",
       "      <td>223</td>\n",
       "      <td>1</td>\n",
       "      <td>normal</td>\n",
       "      <td>male</td>\n",
       "      <td>71-year-old human stage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC3_ACAGCTAAGTGTCCCG</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>233.0</td>\n",
       "      <td>204</td>\n",
       "      <td>1</td>\n",
       "      <td>normal</td>\n",
       "      <td>male</td>\n",
       "      <td>71-year-old human stage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC3_ACAGCTACAAGGACAC</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>285.0</td>\n",
       "      <td>242</td>\n",
       "      <td>1</td>\n",
       "      <td>normal</td>\n",
       "      <td>male</td>\n",
       "      <td>71-year-old human stage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC3_ACAGCTAGTGATGTCT</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>959.0</td>\n",
       "      <td>699</td>\n",
       "      <td>2</td>\n",
       "      <td>normal</td>\n",
       "      <td>male</td>\n",
       "      <td>71-year-old human stage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC10_TTTCCTCGTGGTAACG</th>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>1010.0</td>\n",
       "      <td>710</td>\n",
       "      <td>1</td>\n",
       "      <td>Alzheimer disease</td>\n",
       "      <td>male</td>\n",
       "      <td>82-year-old human stage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC10_TTTCCTCGTTTGACAC</th>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>1355.0</td>\n",
       "      <td>1013</td>\n",
       "      <td>1</td>\n",
       "      <td>Alzheimer disease</td>\n",
       "      <td>male</td>\n",
       "      <td>82-year-old human stage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC10_TTTCCTCTCACAGGCC</th>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>1971.0</td>\n",
       "      <td>1358</td>\n",
       "      <td>0</td>\n",
       "      <td>Alzheimer disease</td>\n",
       "      <td>male</td>\n",
       "      <td>82-year-old human stage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC10_TTTGTCACAAGCGTAG</th>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>403.0</td>\n",
       "      <td>342</td>\n",
       "      <td>2</td>\n",
       "      <td>Alzheimer disease</td>\n",
       "      <td>male</td>\n",
       "      <td>82-year-old human stage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC10_TTTGTCATCTATCGCC</th>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>444.0</td>\n",
       "      <td>393</td>\n",
       "      <td>2</td>\n",
       "      <td>Alzheimer disease</td>\n",
       "      <td>male</td>\n",
       "      <td>82-year-old human stage</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5572 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      donor_id BraakStage    nUMI  nGene seurat.subclusters  \\\n",
       "EC3_AACCATGTCACGAAGG         3          0   219.0    193                  0   \n",
       "EC3_AATCGGTCACTTACGA         3          0   266.0    223                  1   \n",
       "EC3_ACAGCTAAGTGTCCCG         3          0   233.0    204                  1   \n",
       "EC3_ACAGCTACAAGGACAC         3          0   285.0    242                  1   \n",
       "EC3_ACAGCTAGTGATGTCT         3          0   959.0    699                  2   \n",
       "...                        ...        ...     ...    ...                ...   \n",
       "EC10_TTTCCTCGTGGTAACG       10          6  1010.0    710                  1   \n",
       "EC10_TTTCCTCGTTTGACAC       10          6  1355.0   1013                  1   \n",
       "EC10_TTTCCTCTCACAGGCC       10          6  1971.0   1358                  0   \n",
       "EC10_TTTGTCACAAGCGTAG       10          6   403.0    342                  2   \n",
       "EC10_TTTGTCATCTATCGCC       10          6   444.0    393                  2   \n",
       "\n",
       "                                 disease   sex        development_stage  \n",
       "EC3_AACCATGTCACGAAGG              normal  male  71-year-old human stage  \n",
       "EC3_AATCGGTCACTTACGA              normal  male  71-year-old human stage  \n",
       "EC3_ACAGCTAAGTGTCCCG              normal  male  71-year-old human stage  \n",
       "EC3_ACAGCTACAAGGACAC              normal  male  71-year-old human stage  \n",
       "EC3_ACAGCTAGTGATGTCT              normal  male  71-year-old human stage  \n",
       "...                                  ...   ...                      ...  \n",
       "EC10_TTTCCTCGTGGTAACG  Alzheimer disease  male  82-year-old human stage  \n",
       "EC10_TTTCCTCGTTTGACAC  Alzheimer disease  male  82-year-old human stage  \n",
       "EC10_TTTCCTCTCACAGGCC  Alzheimer disease  male  82-year-old human stage  \n",
       "EC10_TTTGTCACAAGCGTAG  Alzheimer disease  male  82-year-old human stage  \n",
       "EC10_TTTGTCATCTATCGCC  Alzheimer disease  male  82-year-old human stage  \n",
       "\n",
       "[5572 rows x 8 columns]"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "e806049c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "View of AnnData object with n_obs × n_vars = 5572 × 24751\n",
       "    obs: 'SampleID', 'donor_id', 'BraakStage', 'SampleBatch', 'nUMI', 'nGene', 'initialClusterAssignments', 'seurat.subclusters', 'subclusterAssignment', 'tissue_ontology_term_id', 'cell_type_ontology_term_id', 'assay_ontology_term_id', 'disease_ontology_term_id', 'self_reported_ethnicity_ontology_term_id', 'development_stage_ontology_term_id', 'sex_ontology_term_id', 'is_primary_data', 'organism_ontology_term_id', 'suspension_type', 'tissue_type', 'cell_type', 'assay', 'disease', 'organism', 'sex', 'tissue', 'self_reported_ethnicity', 'development_stage', 'observation_joinid'\n",
       "    var: 'feature_is_filtered', 'feature_name', 'feature_reference', 'feature_biotype', 'feature_length'\n",
       "    uns: 'citation', 'schema_reference', 'schema_version', 'title'\n",
       "    obsm: 'X_cca', 'X_cca.aligned', 'X_tsne'"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2653febe",
   "metadata": {},
   "outputs": [],
   "source": [
    "#From investigating the metadata dataframe, BraakStage, nUMI, nGene and seurat.subclusters are all numerical, whilst disease, sex and development_stage are all character strings.\n",
    "#The columns with character strings will need to be reformatted appropriately so that they can be correlated against.\n",
    "#Lets first identify the unique labels within each column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "9df567e2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['male']\n",
       "Categories (1, object): ['male']"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata['sex'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b94cf306",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Looks like there are only male participants. Since there are only male differences, this column can also be removed, since we will not be able to investigate sex differences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "f2b6bb83",
   "metadata": {},
   "outputs": [],
   "source": [
    "column_to_remove = 'sex'\n",
    "metadata.drop(columns=column_to_remove, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "35cc1ed4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>donor_id</th>\n",
       "      <th>BraakStage</th>\n",
       "      <th>nUMI</th>\n",
       "      <th>nGene</th>\n",
       "      <th>seurat.subclusters</th>\n",
       "      <th>disease</th>\n",
       "      <th>development_stage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>EC3_AACCATGTCACGAAGG</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>219.0</td>\n",
       "      <td>193</td>\n",
       "      <td>0</td>\n",
       "      <td>normal</td>\n",
       "      <td>71-year-old human stage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC3_AATCGGTCACTTACGA</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>266.0</td>\n",
       "      <td>223</td>\n",
       "      <td>1</td>\n",
       "      <td>normal</td>\n",
       "      <td>71-year-old human stage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC3_ACAGCTAAGTGTCCCG</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>233.0</td>\n",
       "      <td>204</td>\n",
       "      <td>1</td>\n",
       "      <td>normal</td>\n",
       "      <td>71-year-old human stage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC3_ACAGCTACAAGGACAC</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>285.0</td>\n",
       "      <td>242</td>\n",
       "      <td>1</td>\n",
       "      <td>normal</td>\n",
       "      <td>71-year-old human stage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC3_ACAGCTAGTGATGTCT</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>959.0</td>\n",
       "      <td>699</td>\n",
       "      <td>2</td>\n",
       "      <td>normal</td>\n",
       "      <td>71-year-old human stage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC10_TTTCCTCGTGGTAACG</th>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>1010.0</td>\n",
       "      <td>710</td>\n",
       "      <td>1</td>\n",
       "      <td>Alzheimer disease</td>\n",
       "      <td>82-year-old human stage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC10_TTTCCTCGTTTGACAC</th>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>1355.0</td>\n",
       "      <td>1013</td>\n",
       "      <td>1</td>\n",
       "      <td>Alzheimer disease</td>\n",
       "      <td>82-year-old human stage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC10_TTTCCTCTCACAGGCC</th>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>1971.0</td>\n",
       "      <td>1358</td>\n",
       "      <td>0</td>\n",
       "      <td>Alzheimer disease</td>\n",
       "      <td>82-year-old human stage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC10_TTTGTCACAAGCGTAG</th>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>403.0</td>\n",
       "      <td>342</td>\n",
       "      <td>2</td>\n",
       "      <td>Alzheimer disease</td>\n",
       "      <td>82-year-old human stage</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC10_TTTGTCATCTATCGCC</th>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>444.0</td>\n",
       "      <td>393</td>\n",
       "      <td>2</td>\n",
       "      <td>Alzheimer disease</td>\n",
       "      <td>82-year-old human stage</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5572 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      donor_id BraakStage    nUMI  nGene seurat.subclusters  \\\n",
       "EC3_AACCATGTCACGAAGG         3          0   219.0    193                  0   \n",
       "EC3_AATCGGTCACTTACGA         3          0   266.0    223                  1   \n",
       "EC3_ACAGCTAAGTGTCCCG         3          0   233.0    204                  1   \n",
       "EC3_ACAGCTACAAGGACAC         3          0   285.0    242                  1   \n",
       "EC3_ACAGCTAGTGATGTCT         3          0   959.0    699                  2   \n",
       "...                        ...        ...     ...    ...                ...   \n",
       "EC10_TTTCCTCGTGGTAACG       10          6  1010.0    710                  1   \n",
       "EC10_TTTCCTCGTTTGACAC       10          6  1355.0   1013                  1   \n",
       "EC10_TTTCCTCTCACAGGCC       10          6  1971.0   1358                  0   \n",
       "EC10_TTTGTCACAAGCGTAG       10          6   403.0    342                  2   \n",
       "EC10_TTTGTCATCTATCGCC       10          6   444.0    393                  2   \n",
       "\n",
       "                                 disease        development_stage  \n",
       "EC3_AACCATGTCACGAAGG              normal  71-year-old human stage  \n",
       "EC3_AATCGGTCACTTACGA              normal  71-year-old human stage  \n",
       "EC3_ACAGCTAAGTGTCCCG              normal  71-year-old human stage  \n",
       "EC3_ACAGCTACAAGGACAC              normal  71-year-old human stage  \n",
       "EC3_ACAGCTAGTGATGTCT              normal  71-year-old human stage  \n",
       "...                                  ...                      ...  \n",
       "EC10_TTTCCTCGTGGTAACG  Alzheimer disease  82-year-old human stage  \n",
       "EC10_TTTCCTCGTTTGACAC  Alzheimer disease  82-year-old human stage  \n",
       "EC10_TTTCCTCTCACAGGCC  Alzheimer disease  82-year-old human stage  \n",
       "EC10_TTTGTCACAAGCGTAG  Alzheimer disease  82-year-old human stage  \n",
       "EC10_TTTGTCATCTATCGCC  Alzheimer disease  82-year-old human stage  \n",
       "\n",
       "[5572 rows x 7 columns]"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "544073b5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['normal', 'Alzheimer disease']\n",
       "Categories (2, object): ['Alzheimer disease', 'normal']"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Now let's have a look at the disease variable\n",
    "metadata['disease'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "a135f6dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "#The disease column can be encoded into a binary variable:\n",
    "metadata['AD'] = metadata['disease'].apply(lambda x: 1 if x == \"Alzheimer disease\" else 0)\n",
    "metadata['Normal'] = metadata['disease'].apply(lambda x: 1 if x == \"normal\" else 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "99ca55ef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>donor_id</th>\n",
       "      <th>BraakStage</th>\n",
       "      <th>nUMI</th>\n",
       "      <th>nGene</th>\n",
       "      <th>seurat.subclusters</th>\n",
       "      <th>disease</th>\n",
       "      <th>development_stage</th>\n",
       "      <th>AD</th>\n",
       "      <th>Normal</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>EC3_AACCATGTCACGAAGG</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>219.0</td>\n",
       "      <td>193</td>\n",
       "      <td>0</td>\n",
       "      <td>normal</td>\n",
       "      <td>71-year-old human stage</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC3_AATCGGTCACTTACGA</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>266.0</td>\n",
       "      <td>223</td>\n",
       "      <td>1</td>\n",
       "      <td>normal</td>\n",
       "      <td>71-year-old human stage</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC3_ACAGCTAAGTGTCCCG</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>233.0</td>\n",
       "      <td>204</td>\n",
       "      <td>1</td>\n",
       "      <td>normal</td>\n",
       "      <td>71-year-old human stage</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC3_ACAGCTACAAGGACAC</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>285.0</td>\n",
       "      <td>242</td>\n",
       "      <td>1</td>\n",
       "      <td>normal</td>\n",
       "      <td>71-year-old human stage</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC3_ACAGCTAGTGATGTCT</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>959.0</td>\n",
       "      <td>699</td>\n",
       "      <td>2</td>\n",
       "      <td>normal</td>\n",
       "      <td>71-year-old human stage</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC10_TTTCCTCGTGGTAACG</th>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>1010.0</td>\n",
       "      <td>710</td>\n",
       "      <td>1</td>\n",
       "      <td>Alzheimer disease</td>\n",
       "      <td>82-year-old human stage</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC10_TTTCCTCGTTTGACAC</th>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>1355.0</td>\n",
       "      <td>1013</td>\n",
       "      <td>1</td>\n",
       "      <td>Alzheimer disease</td>\n",
       "      <td>82-year-old human stage</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC10_TTTCCTCTCACAGGCC</th>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>1971.0</td>\n",
       "      <td>1358</td>\n",
       "      <td>0</td>\n",
       "      <td>Alzheimer disease</td>\n",
       "      <td>82-year-old human stage</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC10_TTTGTCACAAGCGTAG</th>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>403.0</td>\n",
       "      <td>342</td>\n",
       "      <td>2</td>\n",
       "      <td>Alzheimer disease</td>\n",
       "      <td>82-year-old human stage</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC10_TTTGTCATCTATCGCC</th>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>444.0</td>\n",
       "      <td>393</td>\n",
       "      <td>2</td>\n",
       "      <td>Alzheimer disease</td>\n",
       "      <td>82-year-old human stage</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5572 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      donor_id BraakStage    nUMI  nGene seurat.subclusters  \\\n",
       "EC3_AACCATGTCACGAAGG         3          0   219.0    193                  0   \n",
       "EC3_AATCGGTCACTTACGA         3          0   266.0    223                  1   \n",
       "EC3_ACAGCTAAGTGTCCCG         3          0   233.0    204                  1   \n",
       "EC3_ACAGCTACAAGGACAC         3          0   285.0    242                  1   \n",
       "EC3_ACAGCTAGTGATGTCT         3          0   959.0    699                  2   \n",
       "...                        ...        ...     ...    ...                ...   \n",
       "EC10_TTTCCTCGTGGTAACG       10          6  1010.0    710                  1   \n",
       "EC10_TTTCCTCGTTTGACAC       10          6  1355.0   1013                  1   \n",
       "EC10_TTTCCTCTCACAGGCC       10          6  1971.0   1358                  0   \n",
       "EC10_TTTGTCACAAGCGTAG       10          6   403.0    342                  2   \n",
       "EC10_TTTGTCATCTATCGCC       10          6   444.0    393                  2   \n",
       "\n",
       "                                 disease        development_stage AD Normal  \n",
       "EC3_AACCATGTCACGAAGG              normal  71-year-old human stage  0      1  \n",
       "EC3_AATCGGTCACTTACGA              normal  71-year-old human stage  0      1  \n",
       "EC3_ACAGCTAAGTGTCCCG              normal  71-year-old human stage  0      1  \n",
       "EC3_ACAGCTACAAGGACAC              normal  71-year-old human stage  0      1  \n",
       "EC3_ACAGCTAGTGATGTCT              normal  71-year-old human stage  0      1  \n",
       "...                                  ...                      ... ..    ...  \n",
       "EC10_TTTCCTCGTGGTAACG  Alzheimer disease  82-year-old human stage  1      0  \n",
       "EC10_TTTCCTCGTTTGACAC  Alzheimer disease  82-year-old human stage  1      0  \n",
       "EC10_TTTCCTCTCACAGGCC  Alzheimer disease  82-year-old human stage  1      0  \n",
       "EC10_TTTGTCACAAGCGTAG  Alzheimer disease  82-year-old human stage  1      0  \n",
       "EC10_TTTGTCATCTATCGCC  Alzheimer disease  82-year-old human stage  1      0  \n",
       "\n",
       "[5572 rows x 9 columns]"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48fd095f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Now lets sort out the development_stage column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "6cf7d63b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['71-year-old human stage', '50-year-old human stage', '60-year-old human stage', '77-year-old human stage', '87-year-old human stage', '80 year-old and over human stage', '72-year-old human stage', '82-year-old human stage']\n",
       "Categories (8, object): ['80 year-old and over human stage', '50-year-old human stage', '60-year-old human stage', '71-year-old human stage', '72-year-old human stage', '77-year-old human stage', '82-year-old human stage', '87-year-old human stage']"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata['development_stage'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "79f3a5b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#There appear to be 8 categories. Lets numerically encode them\n",
    "# Recode development_stage\n",
    "development_stage_mapping = {\n",
    "    '50-year-old human stage': 50,\n",
    "    '60-year-old human stage': 60,\n",
    "    '71-year-old human stage': 71,\n",
    "    '72-year-old human stage': 72,\n",
    "    '77-year-old human stage': 77,\n",
    "    '80 year-old and over human stage': 80,\n",
    "    '82-year-old human stage': 82,\n",
    "    '87-year-old human stage': 87\n",
    "}\n",
    "metadata['development_stage'] = metadata['development_stage'].map(development_stage_mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "5e4ac57c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>donor_id</th>\n",
       "      <th>BraakStage</th>\n",
       "      <th>nUMI</th>\n",
       "      <th>nGene</th>\n",
       "      <th>seurat.subclusters</th>\n",
       "      <th>disease</th>\n",
       "      <th>development_stage</th>\n",
       "      <th>AD</th>\n",
       "      <th>Normal</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>EC3_AACCATGTCACGAAGG</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>219.0</td>\n",
       "      <td>193</td>\n",
       "      <td>0</td>\n",
       "      <td>normal</td>\n",
       "      <td>71</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC3_AATCGGTCACTTACGA</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>266.0</td>\n",
       "      <td>223</td>\n",
       "      <td>1</td>\n",
       "      <td>normal</td>\n",
       "      <td>71</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC3_ACAGCTAAGTGTCCCG</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>233.0</td>\n",
       "      <td>204</td>\n",
       "      <td>1</td>\n",
       "      <td>normal</td>\n",
       "      <td>71</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC3_ACAGCTACAAGGACAC</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>285.0</td>\n",
       "      <td>242</td>\n",
       "      <td>1</td>\n",
       "      <td>normal</td>\n",
       "      <td>71</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC3_ACAGCTAGTGATGTCT</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>959.0</td>\n",
       "      <td>699</td>\n",
       "      <td>2</td>\n",
       "      <td>normal</td>\n",
       "      <td>71</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC10_TTTCCTCGTGGTAACG</th>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>1010.0</td>\n",
       "      <td>710</td>\n",
       "      <td>1</td>\n",
       "      <td>Alzheimer disease</td>\n",
       "      <td>82</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC10_TTTCCTCGTTTGACAC</th>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>1355.0</td>\n",
       "      <td>1013</td>\n",
       "      <td>1</td>\n",
       "      <td>Alzheimer disease</td>\n",
       "      <td>82</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC10_TTTCCTCTCACAGGCC</th>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>1971.0</td>\n",
       "      <td>1358</td>\n",
       "      <td>0</td>\n",
       "      <td>Alzheimer disease</td>\n",
       "      <td>82</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC10_TTTGTCACAAGCGTAG</th>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>403.0</td>\n",
       "      <td>342</td>\n",
       "      <td>2</td>\n",
       "      <td>Alzheimer disease</td>\n",
       "      <td>82</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC10_TTTGTCATCTATCGCC</th>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>444.0</td>\n",
       "      <td>393</td>\n",
       "      <td>2</td>\n",
       "      <td>Alzheimer disease</td>\n",
       "      <td>82</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5572 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      donor_id BraakStage    nUMI  nGene seurat.subclusters  \\\n",
       "EC3_AACCATGTCACGAAGG         3          0   219.0    193                  0   \n",
       "EC3_AATCGGTCACTTACGA         3          0   266.0    223                  1   \n",
       "EC3_ACAGCTAAGTGTCCCG         3          0   233.0    204                  1   \n",
       "EC3_ACAGCTACAAGGACAC         3          0   285.0    242                  1   \n",
       "EC3_ACAGCTAGTGATGTCT         3          0   959.0    699                  2   \n",
       "...                        ...        ...     ...    ...                ...   \n",
       "EC10_TTTCCTCGTGGTAACG       10          6  1010.0    710                  1   \n",
       "EC10_TTTCCTCGTTTGACAC       10          6  1355.0   1013                  1   \n",
       "EC10_TTTCCTCTCACAGGCC       10          6  1971.0   1358                  0   \n",
       "EC10_TTTGTCACAAGCGTAG       10          6   403.0    342                  2   \n",
       "EC10_TTTGTCATCTATCGCC       10          6   444.0    393                  2   \n",
       "\n",
       "                                 disease development_stage AD Normal  \n",
       "EC3_AACCATGTCACGAAGG              normal                71  0      1  \n",
       "EC3_AATCGGTCACTTACGA              normal                71  0      1  \n",
       "EC3_ACAGCTAAGTGTCCCG              normal                71  0      1  \n",
       "EC3_ACAGCTACAAGGACAC              normal                71  0      1  \n",
       "EC3_ACAGCTAGTGATGTCT              normal                71  0      1  \n",
       "...                                  ...               ... ..    ...  \n",
       "EC10_TTTCCTCGTGGTAACG  Alzheimer disease                82  1      0  \n",
       "EC10_TTTCCTCGTTTGACAC  Alzheimer disease                82  1      0  \n",
       "EC10_TTTCCTCTCACAGGCC  Alzheimer disease                82  1      0  \n",
       "EC10_TTTGTCACAAGCGTAG  Alzheimer disease                82  1      0  \n",
       "EC10_TTTGTCATCTATCGCC  Alzheimer disease                82  1      0  \n",
       "\n",
       "[5572 rows x 9 columns]"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "7a7ce252",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>donor_id</th>\n",
       "      <th>BraakStage</th>\n",
       "      <th>nUMI</th>\n",
       "      <th>nGene</th>\n",
       "      <th>seurat.subclusters</th>\n",
       "      <th>development_stage</th>\n",
       "      <th>AD</th>\n",
       "      <th>Normal</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>EC3_AACCATGTCACGAAGG</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>219.0</td>\n",
       "      <td>193</td>\n",
       "      <td>0</td>\n",
       "      <td>71</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC3_AATCGGTCACTTACGA</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>266.0</td>\n",
       "      <td>223</td>\n",
       "      <td>1</td>\n",
       "      <td>71</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC3_ACAGCTAAGTGTCCCG</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>233.0</td>\n",
       "      <td>204</td>\n",
       "      <td>1</td>\n",
       "      <td>71</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC3_ACAGCTACAAGGACAC</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>285.0</td>\n",
       "      <td>242</td>\n",
       "      <td>1</td>\n",
       "      <td>71</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC3_ACAGCTAGTGATGTCT</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>959.0</td>\n",
       "      <td>699</td>\n",
       "      <td>2</td>\n",
       "      <td>71</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC10_TTTCCTCGTGGTAACG</th>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>1010.0</td>\n",
       "      <td>710</td>\n",
       "      <td>1</td>\n",
       "      <td>82</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC10_TTTCCTCGTTTGACAC</th>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>1355.0</td>\n",
       "      <td>1013</td>\n",
       "      <td>1</td>\n",
       "      <td>82</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC10_TTTCCTCTCACAGGCC</th>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>1971.0</td>\n",
       "      <td>1358</td>\n",
       "      <td>0</td>\n",
       "      <td>82</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC10_TTTGTCACAAGCGTAG</th>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>403.0</td>\n",
       "      <td>342</td>\n",
       "      <td>2</td>\n",
       "      <td>82</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC10_TTTGTCATCTATCGCC</th>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>444.0</td>\n",
       "      <td>393</td>\n",
       "      <td>2</td>\n",
       "      <td>82</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5572 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      donor_id BraakStage    nUMI  nGene seurat.subclusters  \\\n",
       "EC3_AACCATGTCACGAAGG         3          0   219.0    193                  0   \n",
       "EC3_AATCGGTCACTTACGA         3          0   266.0    223                  1   \n",
       "EC3_ACAGCTAAGTGTCCCG         3          0   233.0    204                  1   \n",
       "EC3_ACAGCTACAAGGACAC         3          0   285.0    242                  1   \n",
       "EC3_ACAGCTAGTGATGTCT         3          0   959.0    699                  2   \n",
       "...                        ...        ...     ...    ...                ...   \n",
       "EC10_TTTCCTCGTGGTAACG       10          6  1010.0    710                  1   \n",
       "EC10_TTTCCTCGTTTGACAC       10          6  1355.0   1013                  1   \n",
       "EC10_TTTCCTCTCACAGGCC       10          6  1971.0   1358                  0   \n",
       "EC10_TTTGTCACAAGCGTAG       10          6   403.0    342                  2   \n",
       "EC10_TTTGTCATCTATCGCC       10          6   444.0    393                  2   \n",
       "\n",
       "                      development_stage AD Normal  \n",
       "EC3_AACCATGTCACGAAGG                 71  0      1  \n",
       "EC3_AATCGGTCACTTACGA                 71  0      1  \n",
       "EC3_ACAGCTAAGTGTCCCG                 71  0      1  \n",
       "EC3_ACAGCTACAAGGACAC                 71  0      1  \n",
       "EC3_ACAGCTAGTGATGTCT                 71  0      1  \n",
       "...                                 ... ..    ...  \n",
       "EC10_TTTCCTCGTGGTAACG                82  1      0  \n",
       "EC10_TTTCCTCGTTTGACAC                82  1      0  \n",
       "EC10_TTTCCTCTCACAGGCC                82  1      0  \n",
       "EC10_TTTGTCACAAGCGTAG                82  1      0  \n",
       "EC10_TTTGTCATCTATCGCC                82  1      0  \n",
       "\n",
       "[5572 rows x 8 columns]"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Drop the disease column as it is no longer necessary\n",
    "# Drop unnecessary columns\n",
    "metadata = metadata.drop(['disease'], axis=1)\n",
    "metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "27007b63",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Save the metadata dataframe\n",
    "metadata.to_csv('data/mic_metadata.csv', index = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "9579f0eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata = pd.read_csv('data/mic_metadata.csv', index_col = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5af7e749",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Due to the nature of single-cell data, we naturally have many cells from the same donor.\n",
    "#However, we cannot simply correlate the gene expression data in its current form. this would lead to within and outwith donor correlations.\n",
    "#Therefore, since we are working with single-cell data, this must first be pseudobulked in order to continue with the analysis.\n",
    "#This is important as it not only speeds up the computation, but most importantly negates the effects of within sample correlation.\n",
    "#Also, pseudobulking can help to mitigate the issues commonly found in single-cell data, such as drop outs and high zero value counts."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56513fc2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8a7337a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#First we shall sort out the metadata dataframe so that it only contains one row per donor since the data will be aggregated."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9794446f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "e6007dcc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert row names to a column named 'cell_id'\n",
    "metadata['cell_id'] = metadata.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "36bcec4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Group by 'donor_id' and select the first row of each group\n",
    "rows = metadata.groupby('donor_id').first().reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "ee17431d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>donor_id</th>\n",
       "      <th>BraakStage</th>\n",
       "      <th>nUMI</th>\n",
       "      <th>nGene</th>\n",
       "      <th>seurat.subclusters</th>\n",
       "      <th>development_stage</th>\n",
       "      <th>AD</th>\n",
       "      <th>Normal</th>\n",
       "      <th>cell_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>302.0</td>\n",
       "      <td>268</td>\n",
       "      <td>1</td>\n",
       "      <td>50</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>EC1_AAAGATGAGGAGTTTA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>676.0</td>\n",
       "      <td>553</td>\n",
       "      <td>1</td>\n",
       "      <td>60</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>EC2_AAAGCAAAGCTACCGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>219.0</td>\n",
       "      <td>193</td>\n",
       "      <td>0</td>\n",
       "      <td>71</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>EC3_AACCATGTCACGAAGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>348.0</td>\n",
       "      <td>295</td>\n",
       "      <td>0</td>\n",
       "      <td>72</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>EC4_AAAGCAAGTATGGTTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>447.0</td>\n",
       "      <td>328</td>\n",
       "      <td>1</td>\n",
       "      <td>77</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>EC5_AAACCTGAGGCGTACA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>376.0</td>\n",
       "      <td>328</td>\n",
       "      <td>2</td>\n",
       "      <td>87</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>EC6_AAACGGGTCGAGAGCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>2</td>\n",
       "      <td>1399.0</td>\n",
       "      <td>965</td>\n",
       "      <td>1</td>\n",
       "      <td>80</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>EC7_AAACCTGGTCGTTGTA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>8</td>\n",
       "      <td>6</td>\n",
       "      <td>619.0</td>\n",
       "      <td>500</td>\n",
       "      <td>0</td>\n",
       "      <td>72</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>EC8_AAACCTGAGCCGCCTA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>9</td>\n",
       "      <td>6</td>\n",
       "      <td>531.0</td>\n",
       "      <td>437</td>\n",
       "      <td>0</td>\n",
       "      <td>82</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>EC9_AAACCTGCAGCTTCGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>688.0</td>\n",
       "      <td>549</td>\n",
       "      <td>1</td>\n",
       "      <td>82</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>EC10_AAACGGGGTTGGAGGT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  donor_id BraakStage    nUMI  nGene seurat.subclusters development_stage AD  \\\n",
       "0        1          0   302.0    268                  1                50  0   \n",
       "1        2          0   676.0    553                  1                60  0   \n",
       "2        3          0   219.0    193                  0                71  0   \n",
       "3        4          2   348.0    295                  0                72  1   \n",
       "4        5          2   447.0    328                  1                77  1   \n",
       "5        6          2   376.0    328                  2                87  1   \n",
       "6        7          2  1399.0    965                  1                80  1   \n",
       "7        8          6   619.0    500                  0                72  1   \n",
       "8        9          6   531.0    437                  0                82  1   \n",
       "9       10          6   688.0    549                  1                82  1   \n",
       "\n",
       "  Normal                cell_id  \n",
       "0      1   EC1_AAAGATGAGGAGTTTA  \n",
       "1      1   EC2_AAAGCAAAGCTACCGC  \n",
       "2      1   EC3_AACCATGTCACGAAGG  \n",
       "3      0   EC4_AAAGCAAGTATGGTTC  \n",
       "4      0   EC5_AAACCTGAGGCGTACA  \n",
       "5      0   EC6_AAACGGGTCGAGAGCA  \n",
       "6      0   EC7_AAACCTGGTCGTTGTA  \n",
       "7      0   EC8_AAACCTGAGCCGCCTA  \n",
       "8      0   EC9_AAACCTGCAGCTTCGG  \n",
       "9      0  EC10_AAACGGGGTTGGAGGT  "
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "12129a1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract row indices corresponding to the first cell from each donor\n",
    "row_list = []\n",
    "for i, row in rows.iterrows():\n",
    "    row_idx = metadata.index.get_loc(row['cell_id'])\n",
    "    row_list.append(row_idx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "3aa3e79f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[70, 318, 0, 1675, 620, 856, 1140, 1777, 3114, 4812]"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "row_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "b44673b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Select the columns from the DataFrame\n",
    "metadata2 = metadata.iloc[row_list, :].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "9de3d9e6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>donor_id</th>\n",
       "      <th>BraakStage</th>\n",
       "      <th>nUMI</th>\n",
       "      <th>nGene</th>\n",
       "      <th>seurat.subclusters</th>\n",
       "      <th>development_stage</th>\n",
       "      <th>AD</th>\n",
       "      <th>Normal</th>\n",
       "      <th>cell_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>EC1_AAAGATGAGGAGTTTA</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>302.0</td>\n",
       "      <td>268</td>\n",
       "      <td>1</td>\n",
       "      <td>50</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>EC1_AAAGATGAGGAGTTTA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC2_AAAGCAAAGCTACCGC</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>676.0</td>\n",
       "      <td>553</td>\n",
       "      <td>1</td>\n",
       "      <td>60</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>EC2_AAAGCAAAGCTACCGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC3_AACCATGTCACGAAGG</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>219.0</td>\n",
       "      <td>193</td>\n",
       "      <td>0</td>\n",
       "      <td>71</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>EC3_AACCATGTCACGAAGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC4_AAAGCAAGTATGGTTC</th>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>348.0</td>\n",
       "      <td>295</td>\n",
       "      <td>0</td>\n",
       "      <td>72</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>EC4_AAAGCAAGTATGGTTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC5_AAACCTGAGGCGTACA</th>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>447.0</td>\n",
       "      <td>328</td>\n",
       "      <td>1</td>\n",
       "      <td>77</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>EC5_AAACCTGAGGCGTACA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC6_AAACGGGTCGAGAGCA</th>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>376.0</td>\n",
       "      <td>328</td>\n",
       "      <td>2</td>\n",
       "      <td>87</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>EC6_AAACGGGTCGAGAGCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC7_AAACCTGGTCGTTGTA</th>\n",
       "      <td>7</td>\n",
       "      <td>2</td>\n",
       "      <td>1399.0</td>\n",
       "      <td>965</td>\n",
       "      <td>1</td>\n",
       "      <td>80</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>EC7_AAACCTGGTCGTTGTA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC8_AAACCTGAGCCGCCTA</th>\n",
       "      <td>8</td>\n",
       "      <td>6</td>\n",
       "      <td>619.0</td>\n",
       "      <td>500</td>\n",
       "      <td>0</td>\n",
       "      <td>72</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>EC8_AAACCTGAGCCGCCTA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC9_AAACCTGCAGCTTCGG</th>\n",
       "      <td>9</td>\n",
       "      <td>6</td>\n",
       "      <td>531.0</td>\n",
       "      <td>437</td>\n",
       "      <td>0</td>\n",
       "      <td>82</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>EC9_AAACCTGCAGCTTCGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC10_AAACGGGGTTGGAGGT</th>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>688.0</td>\n",
       "      <td>549</td>\n",
       "      <td>1</td>\n",
       "      <td>82</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>EC10_AAACGGGGTTGGAGGT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      donor_id BraakStage    nUMI  nGene seurat.subclusters  \\\n",
       "EC1_AAAGATGAGGAGTTTA         1          0   302.0    268                  1   \n",
       "EC2_AAAGCAAAGCTACCGC         2          0   676.0    553                  1   \n",
       "EC3_AACCATGTCACGAAGG         3          0   219.0    193                  0   \n",
       "EC4_AAAGCAAGTATGGTTC         4          2   348.0    295                  0   \n",
       "EC5_AAACCTGAGGCGTACA         5          2   447.0    328                  1   \n",
       "EC6_AAACGGGTCGAGAGCA         6          2   376.0    328                  2   \n",
       "EC7_AAACCTGGTCGTTGTA         7          2  1399.0    965                  1   \n",
       "EC8_AAACCTGAGCCGCCTA         8          6   619.0    500                  0   \n",
       "EC9_AAACCTGCAGCTTCGG         9          6   531.0    437                  0   \n",
       "EC10_AAACGGGGTTGGAGGT       10          6   688.0    549                  1   \n",
       "\n",
       "                      development_stage AD Normal                cell_id  \n",
       "EC1_AAAGATGAGGAGTTTA                 50  0      1   EC1_AAAGATGAGGAGTTTA  \n",
       "EC2_AAAGCAAAGCTACCGC                 60  0      1   EC2_AAAGCAAAGCTACCGC  \n",
       "EC3_AACCATGTCACGAAGG                 71  0      1   EC3_AACCATGTCACGAAGG  \n",
       "EC4_AAAGCAAGTATGGTTC                 72  1      0   EC4_AAAGCAAGTATGGTTC  \n",
       "EC5_AAACCTGAGGCGTACA                 77  1      0   EC5_AAACCTGAGGCGTACA  \n",
       "EC6_AAACGGGTCGAGAGCA                 87  1      0   EC6_AAACGGGTCGAGAGCA  \n",
       "EC7_AAACCTGGTCGTTGTA                 80  1      0   EC7_AAACCTGGTCGTTGTA  \n",
       "EC8_AAACCTGAGCCGCCTA                 72  1      0   EC8_AAACCTGAGCCGCCTA  \n",
       "EC9_AAACCTGCAGCTTCGG                 82  1      0   EC9_AAACCTGCAGCTTCGG  \n",
       "EC10_AAACGGGGTTGGAGGT                82  1      0  EC10_AAACGGGGTTGGAGGT  "
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "e38fc126",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata2.set_index('donor_id', inplace = True, drop = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "bce9a04b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>donor_id</th>\n",
       "      <th>BraakStage</th>\n",
       "      <th>nUMI</th>\n",
       "      <th>nGene</th>\n",
       "      <th>seurat.subclusters</th>\n",
       "      <th>development_stage</th>\n",
       "      <th>AD</th>\n",
       "      <th>Normal</th>\n",
       "      <th>cell_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>donor_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>302.0</td>\n",
       "      <td>268</td>\n",
       "      <td>1</td>\n",
       "      <td>50</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>EC1_AAAGATGAGGAGTTTA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>676.0</td>\n",
       "      <td>553</td>\n",
       "      <td>1</td>\n",
       "      <td>60</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>EC2_AAAGCAAAGCTACCGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>219.0</td>\n",
       "      <td>193</td>\n",
       "      <td>0</td>\n",
       "      <td>71</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>EC3_AACCATGTCACGAAGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>348.0</td>\n",
       "      <td>295</td>\n",
       "      <td>0</td>\n",
       "      <td>72</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>EC4_AAAGCAAGTATGGTTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>447.0</td>\n",
       "      <td>328</td>\n",
       "      <td>1</td>\n",
       "      <td>77</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>EC5_AAACCTGAGGCGTACA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>376.0</td>\n",
       "      <td>328</td>\n",
       "      <td>2</td>\n",
       "      <td>87</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>EC6_AAACGGGTCGAGAGCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>2</td>\n",
       "      <td>1399.0</td>\n",
       "      <td>965</td>\n",
       "      <td>1</td>\n",
       "      <td>80</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>EC7_AAACCTGGTCGTTGTA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8</td>\n",
       "      <td>6</td>\n",
       "      <td>619.0</td>\n",
       "      <td>500</td>\n",
       "      <td>0</td>\n",
       "      <td>72</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>EC8_AAACCTGAGCCGCCTA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>6</td>\n",
       "      <td>531.0</td>\n",
       "      <td>437</td>\n",
       "      <td>0</td>\n",
       "      <td>82</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>EC9_AAACCTGCAGCTTCGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>688.0</td>\n",
       "      <td>549</td>\n",
       "      <td>1</td>\n",
       "      <td>82</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>EC10_AAACGGGGTTGGAGGT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         donor_id BraakStage    nUMI  nGene seurat.subclusters  \\\n",
       "donor_id                                                         \n",
       "1               1          0   302.0    268                  1   \n",
       "2               2          0   676.0    553                  1   \n",
       "3               3          0   219.0    193                  0   \n",
       "4               4          2   348.0    295                  0   \n",
       "5               5          2   447.0    328                  1   \n",
       "6               6          2   376.0    328                  2   \n",
       "7               7          2  1399.0    965                  1   \n",
       "8               8          6   619.0    500                  0   \n",
       "9               9          6   531.0    437                  0   \n",
       "10             10          6   688.0    549                  1   \n",
       "\n",
       "         development_stage AD Normal                cell_id  \n",
       "donor_id                                                     \n",
       "1                       50  0      1   EC1_AAAGATGAGGAGTTTA  \n",
       "2                       60  0      1   EC2_AAAGCAAAGCTACCGC  \n",
       "3                       71  0      1   EC3_AACCATGTCACGAAGG  \n",
       "4                       72  1      0   EC4_AAAGCAAGTATGGTTC  \n",
       "5                       77  1      0   EC5_AAACCTGAGGCGTACA  \n",
       "6                       87  1      0   EC6_AAACGGGTCGAGAGCA  \n",
       "7                       80  1      0   EC7_AAACCTGGTCGTTGTA  \n",
       "8                       72  1      0   EC8_AAACCTGAGCCGCCTA  \n",
       "9                       82  1      0   EC9_AAACCTGCAGCTTCGG  \n",
       "10                      82  1      0  EC10_AAACGGGGTTGGAGGT  "
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "9cc9b964",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Remove the cell_id column\n",
    "metadata2.drop(columns = 'cell_id', inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "id": "854890da",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>donor_id</th>\n",
       "      <th>BraakStage</th>\n",
       "      <th>nUMI</th>\n",
       "      <th>nGene</th>\n",
       "      <th>seurat.subclusters</th>\n",
       "      <th>development_stage</th>\n",
       "      <th>AD</th>\n",
       "      <th>Normal</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>donor_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>302.0</td>\n",
       "      <td>268</td>\n",
       "      <td>1</td>\n",
       "      <td>50</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>676.0</td>\n",
       "      <td>553</td>\n",
       "      <td>1</td>\n",
       "      <td>60</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>219.0</td>\n",
       "      <td>193</td>\n",
       "      <td>0</td>\n",
       "      <td>71</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>348.0</td>\n",
       "      <td>295</td>\n",
       "      <td>0</td>\n",
       "      <td>72</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>447.0</td>\n",
       "      <td>328</td>\n",
       "      <td>1</td>\n",
       "      <td>77</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>376.0</td>\n",
       "      <td>328</td>\n",
       "      <td>2</td>\n",
       "      <td>87</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>2</td>\n",
       "      <td>1399.0</td>\n",
       "      <td>965</td>\n",
       "      <td>1</td>\n",
       "      <td>80</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8</td>\n",
       "      <td>6</td>\n",
       "      <td>619.0</td>\n",
       "      <td>500</td>\n",
       "      <td>0</td>\n",
       "      <td>72</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>6</td>\n",
       "      <td>531.0</td>\n",
       "      <td>437</td>\n",
       "      <td>0</td>\n",
       "      <td>82</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>688.0</td>\n",
       "      <td>549</td>\n",
       "      <td>1</td>\n",
       "      <td>82</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         donor_id BraakStage    nUMI  nGene seurat.subclusters  \\\n",
       "donor_id                                                         \n",
       "1               1          0   302.0    268                  1   \n",
       "2               2          0   676.0    553                  1   \n",
       "3               3          0   219.0    193                  0   \n",
       "4               4          2   348.0    295                  0   \n",
       "5               5          2   447.0    328                  1   \n",
       "6               6          2   376.0    328                  2   \n",
       "7               7          2  1399.0    965                  1   \n",
       "8               8          6   619.0    500                  0   \n",
       "9               9          6   531.0    437                  0   \n",
       "10             10          6   688.0    549                  1   \n",
       "\n",
       "         development_stage AD Normal  \n",
       "donor_id                              \n",
       "1                       50  0      1  \n",
       "2                       60  0      1  \n",
       "3                       71  0      1  \n",
       "4                       72  1      0  \n",
       "5                       77  1      0  \n",
       "6                       87  1      0  \n",
       "7                       80  1      0  \n",
       "8                       72  1      0  \n",
       "9                       82  1      0  \n",
       "10                      82  1      0  "
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "id": "28fe74f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Save the metadata\n",
    "metadata2.to_csv('data//mic_metadata_pseudobulk.csv', index = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "643000c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#The metadata dataframe for the pseudobulk is now complete"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dda64571",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Lets proceed to aggregate the gene expression data.\n",
    "#This involves summing the gene expression data for each gene of each donor."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55cb95a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#First the gene expression matrix will need to be extracted from our mic adata object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "782322af",
   "metadata": {},
   "outputs": [],
   "source": [
    "#since we are working with single-cell data which will be stored as a sparse matrix, this must be coerced into a dense matrix, so that it can be converted to a dataframe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ec3d90e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert the sparse matrix to a dense matrix\n",
    "dense_matrix = mic.X.todense()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c7539c6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "datExpr = pd.DataFrame(dense_matrix, index=mic.obs_names, columns=mic.var_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "1881d7f9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>feature_name</th>\n",
       "      <th>VTI1B</th>\n",
       "      <th>ASIP</th>\n",
       "      <th>HLX</th>\n",
       "      <th>C16orf87</th>\n",
       "      <th>HMBOX1</th>\n",
       "      <th>RECQL5</th>\n",
       "      <th>NXPH1</th>\n",
       "      <th>SLC5A10</th>\n",
       "      <th>VN1R2</th>\n",
       "      <th>PTOV1</th>\n",
       "      <th>...</th>\n",
       "      <th>SIX3</th>\n",
       "      <th>ACP1</th>\n",
       "      <th>TRBV29OR9-2</th>\n",
       "      <th>OR13A1</th>\n",
       "      <th>LINC02715</th>\n",
       "      <th>RNF149</th>\n",
       "      <th>NRBP2</th>\n",
       "      <th>FOCAD</th>\n",
       "      <th>MGST1</th>\n",
       "      <th>RAB5C-AS1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>EC3_AACCATGTCACGAAGG</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC3_AATCGGTCACTTACGA</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC3_ACAGCTAAGTGTCCCG</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC3_ACAGCTACAAGGACAC</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC3_ACAGCTAGTGATGTCT</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC10_TTTCCTCGTGGTAACG</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC10_TTTCCTCGTTTGACAC</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC10_TTTCCTCTCACAGGCC</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC10_TTTGTCACAAGCGTAG</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC10_TTTGTCATCTATCGCC</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5572 rows × 24751 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "feature_name           VTI1B  ASIP  HLX  C16orf87    HMBOX1  RECQL5     NXPH1  \\\n",
       "EC3_AACCATGTCACGAAGG     0.0   0.0  0.0       0.0  0.000000     0.0  0.000000   \n",
       "EC3_AATCGGTCACTTACGA     0.0   0.0  0.0       0.0  0.000000     0.0  0.000000   \n",
       "EC3_ACAGCTAAGTGTCCCG     0.0   0.0  0.0       0.0  0.000000     0.0  0.000000   \n",
       "EC3_ACAGCTACAAGGACAC     0.0   0.0  0.0       0.0  0.000000     0.0  0.000000   \n",
       "EC3_ACAGCTAGTGATGTCT     0.0   0.0  0.0       0.0  0.000000     0.0  0.693147   \n",
       "...                      ...   ...  ...       ...       ...     ...       ...   \n",
       "EC10_TTTCCTCGTGGTAACG    0.0   0.0  0.0       0.0  0.000000     0.0  0.000000   \n",
       "EC10_TTTCCTCGTTTGACAC    0.0   0.0  0.0       0.0  0.693147     0.0  0.000000   \n",
       "EC10_TTTCCTCTCACAGGCC    0.0   0.0  0.0       0.0  0.000000     0.0  0.000000   \n",
       "EC10_TTTGTCACAAGCGTAG    0.0   0.0  0.0       0.0  0.000000     0.0  0.000000   \n",
       "EC10_TTTGTCATCTATCGCC    0.0   0.0  0.0       0.0  0.000000     0.0  0.000000   \n",
       "\n",
       "feature_name           SLC5A10  VN1R2  PTOV1  ...  SIX3  ACP1  TRBV29OR9-2  \\\n",
       "EC3_AACCATGTCACGAAGG       0.0    0.0    0.0  ...   0.0   0.0          0.0   \n",
       "EC3_AATCGGTCACTTACGA       0.0    0.0    0.0  ...   0.0   0.0          0.0   \n",
       "EC3_ACAGCTAAGTGTCCCG       0.0    0.0    0.0  ...   0.0   0.0          0.0   \n",
       "EC3_ACAGCTACAAGGACAC       0.0    0.0    0.0  ...   0.0   0.0          0.0   \n",
       "EC3_ACAGCTAGTGATGTCT       0.0    0.0    0.0  ...   0.0   0.0          0.0   \n",
       "...                        ...    ...    ...  ...   ...   ...          ...   \n",
       "EC10_TTTCCTCGTGGTAACG      0.0    0.0    0.0  ...   0.0   0.0          0.0   \n",
       "EC10_TTTCCTCGTTTGACAC      0.0    0.0    0.0  ...   0.0   0.0          0.0   \n",
       "EC10_TTTCCTCTCACAGGCC      0.0    0.0    0.0  ...   0.0   0.0          0.0   \n",
       "EC10_TTTGTCACAAGCGTAG      0.0    0.0    0.0  ...   0.0   0.0          0.0   \n",
       "EC10_TTTGTCATCTATCGCC      0.0    0.0    0.0  ...   0.0   0.0          0.0   \n",
       "\n",
       "feature_name           OR13A1  LINC02715    RNF149     NRBP2  FOCAD  MGST1  \\\n",
       "EC3_AACCATGTCACGAAGG      0.0        0.0  0.693147  0.000000    0.0    0.0   \n",
       "EC3_AATCGGTCACTTACGA      0.0        0.0  0.000000  0.693147    0.0    0.0   \n",
       "EC3_ACAGCTAAGTGTCCCG      0.0        0.0  0.000000  0.000000    0.0    0.0   \n",
       "EC3_ACAGCTACAAGGACAC      0.0        0.0  0.000000  0.000000    0.0    0.0   \n",
       "EC3_ACAGCTAGTGATGTCT      0.0        0.0  0.000000  0.000000    0.0    0.0   \n",
       "...                       ...        ...       ...       ...    ...    ...   \n",
       "EC10_TTTCCTCGTGGTAACG     0.0        0.0  0.693147  0.000000    0.0    0.0   \n",
       "EC10_TTTCCTCGTTTGACAC     0.0        0.0  0.000000  0.000000    0.0    0.0   \n",
       "EC10_TTTCCTCTCACAGGCC     0.0        0.0  0.000000  0.000000    0.0    0.0   \n",
       "EC10_TTTGTCACAAGCGTAG     0.0        0.0  0.693147  0.000000    0.0    0.0   \n",
       "EC10_TTTGTCATCTATCGCC     0.0        0.0  0.000000  0.000000    0.0    0.0   \n",
       "\n",
       "feature_name           RAB5C-AS1  \n",
       "EC3_AACCATGTCACGAAGG         0.0  \n",
       "EC3_AATCGGTCACTTACGA         0.0  \n",
       "EC3_ACAGCTAAGTGTCCCG         0.0  \n",
       "EC3_ACAGCTACAAGGACAC         0.0  \n",
       "EC3_ACAGCTAGTGATGTCT         0.0  \n",
       "...                          ...  \n",
       "EC10_TTTCCTCGTGGTAACG        0.0  \n",
       "EC10_TTTCCTCGTTTGACAC        0.0  \n",
       "EC10_TTTCCTCTCACAGGCC        0.0  \n",
       "EC10_TTTGTCACAAGCGTAG        0.0  \n",
       "EC10_TTTGTCATCTATCGCC        0.0  \n",
       "\n",
       "[5572 rows x 24751 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "datExpr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9952a59d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#save datExpr\n",
    "#Save the metadata dataframe\n",
    "datExpr.to_csv('dataset/mic_datExpr_singlecell.csv', index = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "359329b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Since highly variable genes capture the most informative genes, they will be used to filter the expression matrix further.\n",
    "#This is also a way to reduce the dimensionality of the data, so that downstream analyses may be more computationally efficient."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "623c1879",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['UBE2D2', 'SAR1A', 'IER3', 'RASGEF1C', 'TTC7A', 'TPK1', 'PTGES3',\n",
       "       'HMG20B', 'GEM', 'TREM1',\n",
       "       ...\n",
       "       'STAB1', 'NFAT5', 'MAN1A2', 'EDA', 'SNCA', 'FAM135A', 'H2BC21',\n",
       "       'FCGR3A', 'ARL2-SNX15', 'VSIG4'],\n",
       "      dtype='object', name='feature_name', length=999)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hvg = mic.var_names[mic.var['highly_variable']]\n",
    "hvg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "0b5e4db6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>feature_name</th>\n",
       "      <th>UBE2D2</th>\n",
       "      <th>SAR1A</th>\n",
       "      <th>IER3</th>\n",
       "      <th>RASGEF1C</th>\n",
       "      <th>TTC7A</th>\n",
       "      <th>TPK1</th>\n",
       "      <th>PTGES3</th>\n",
       "      <th>HMG20B</th>\n",
       "      <th>GEM</th>\n",
       "      <th>TREM1</th>\n",
       "      <th>...</th>\n",
       "      <th>STAB1</th>\n",
       "      <th>NFAT5</th>\n",
       "      <th>MAN1A2</th>\n",
       "      <th>EDA</th>\n",
       "      <th>SNCA</th>\n",
       "      <th>FAM135A</th>\n",
       "      <th>H2BC21</th>\n",
       "      <th>FCGR3A</th>\n",
       "      <th>ARL2-SNX15</th>\n",
       "      <th>VSIG4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>EC3_AACCATGTCACGAAGG</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC3_AATCGGTCACTTACGA</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC3_ACAGCTAAGTGTCCCG</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC3_ACAGCTACAAGGACAC</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC3_ACAGCTAGTGATGTCT</th>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC10_TTTCCTCGTGGTAACG</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC10_TTTCCTCGTTTGACAC</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC10_TTTCCTCTCACAGGCC</th>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.693147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC10_TTTGTCACAAGCGTAG</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC10_TTTGTCATCTATCGCC</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.098612</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5572 rows × 999 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "feature_name             UBE2D2  SAR1A      IER3  RASGEF1C  TTC7A      TPK1  \\\n",
       "EC3_AACCATGTCACGAAGG   0.000000    0.0  0.000000  0.000000    0.0  0.000000   \n",
       "EC3_AATCGGTCACTTACGA   0.000000    0.0  0.000000  0.000000    0.0  0.000000   \n",
       "EC3_ACAGCTAAGTGTCCCG   0.000000    0.0  0.000000  0.000000    0.0  0.000000   \n",
       "EC3_ACAGCTACAAGGACAC   0.000000    0.0  0.000000  0.000000    0.0  0.000000   \n",
       "EC3_ACAGCTAGTGATGTCT   0.693147    0.0  0.693147  0.000000    0.0  0.000000   \n",
       "...                         ...    ...       ...       ...    ...       ...   \n",
       "EC10_TTTCCTCGTGGTAACG  0.000000    0.0  0.000000  0.000000    0.0  0.000000   \n",
       "EC10_TTTCCTCGTTTGACAC  0.000000    0.0  0.000000  0.000000    0.0  0.000000   \n",
       "EC10_TTTCCTCTCACAGGCC  0.693147    0.0  0.000000  0.693147    0.0  0.693147   \n",
       "EC10_TTTGTCACAAGCGTAG  0.000000    0.0  0.000000  0.000000    0.0  0.000000   \n",
       "EC10_TTTGTCATCTATCGCC  0.000000    0.0  0.000000  0.000000    0.0  0.000000   \n",
       "\n",
       "feature_name           PTGES3  HMG20B  GEM  TREM1  ...     STAB1     NFAT5  \\\n",
       "EC3_AACCATGTCACGAAGG      0.0     0.0  0.0    0.0  ...  0.000000  0.000000   \n",
       "EC3_AATCGGTCACTTACGA      0.0     0.0  0.0    0.0  ...  0.000000  0.000000   \n",
       "EC3_ACAGCTAAGTGTCCCG      0.0     0.0  0.0    0.0  ...  0.693147  0.000000   \n",
       "EC3_ACAGCTACAAGGACAC      0.0     0.0  0.0    0.0  ...  0.000000  0.000000   \n",
       "EC3_ACAGCTAGTGATGTCT      0.0     0.0  0.0    0.0  ...  0.000000  0.000000   \n",
       "...                       ...     ...  ...    ...  ...       ...       ...   \n",
       "EC10_TTTCCTCGTGGTAACG     0.0     0.0  0.0    0.0  ...  0.000000  0.000000   \n",
       "EC10_TTTCCTCGTTTGACAC     0.0     0.0  0.0    0.0  ...  0.693147  0.000000   \n",
       "EC10_TTTCCTCTCACAGGCC     0.0     0.0  0.0    0.0  ...  0.000000  0.693147   \n",
       "EC10_TTTGTCACAAGCGTAG     0.0     0.0  0.0    0.0  ...  0.000000  0.000000   \n",
       "EC10_TTTGTCATCTATCGCC     0.0     0.0  0.0    0.0  ...  0.000000  0.000000   \n",
       "\n",
       "feature_name           MAN1A2  EDA  SNCA   FAM135A  H2BC21  FCGR3A  \\\n",
       "EC3_AACCATGTCACGAAGG      0.0  0.0   0.0  0.000000     0.0     0.0   \n",
       "EC3_AATCGGTCACTTACGA      0.0  0.0   0.0  0.000000     0.0     0.0   \n",
       "EC3_ACAGCTAAGTGTCCCG      0.0  0.0   0.0  0.000000     0.0     0.0   \n",
       "EC3_ACAGCTACAAGGACAC      0.0  0.0   0.0  0.000000     0.0     0.0   \n",
       "EC3_ACAGCTAGTGATGTCT      0.0  0.0   0.0  0.000000     0.0     0.0   \n",
       "...                       ...  ...   ...       ...     ...     ...   \n",
       "EC10_TTTCCTCGTGGTAACG     0.0  0.0   0.0  0.000000     0.0     0.0   \n",
       "EC10_TTTCCTCGTTTGACAC     0.0  0.0   0.0  0.693147     0.0     0.0   \n",
       "EC10_TTTCCTCTCACAGGCC     0.0  0.0   0.0  0.000000     0.0     0.0   \n",
       "EC10_TTTGTCACAAGCGTAG     0.0  0.0   0.0  0.693147     0.0     0.0   \n",
       "EC10_TTTGTCATCTATCGCC     0.0  0.0   0.0  1.098612     0.0     0.0   \n",
       "\n",
       "feature_name           ARL2-SNX15     VSIG4  \n",
       "EC3_AACCATGTCACGAAGG     0.000000  0.000000  \n",
       "EC3_AATCGGTCACTTACGA     0.000000  0.000000  \n",
       "EC3_ACAGCTAAGTGTCCCG     0.000000  0.000000  \n",
       "EC3_ACAGCTACAAGGACAC     0.000000  0.000000  \n",
       "EC3_ACAGCTAGTGATGTCT     0.000000  0.000000  \n",
       "...                           ...       ...  \n",
       "EC10_TTTCCTCGTGGTAACG    0.000000  0.000000  \n",
       "EC10_TTTCCTCGTTTGACAC    0.000000  0.000000  \n",
       "EC10_TTTCCTCTCACAGGCC    0.693147  0.693147  \n",
       "EC10_TTTGTCACAAGCGTAG    0.000000  0.000000  \n",
       "EC10_TTTGTCATCTATCGCC    0.000000  0.000000  \n",
       "\n",
       "[5572 rows x 999 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "datExpr = datExpr.loc[:,hvg]\n",
    "datExpr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "50c46584",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "367653a5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3dc95152",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Add the donor_id column to the gene expression dataframe, so we know which cell came from which donor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "38221089",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reset the index of 'datExpr' DataFrame to make the row names (cell names) a column\n",
    "datExpr_donor = datExpr.reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "427f2e44",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>feature_name</th>\n",
       "      <th>index</th>\n",
       "      <th>UBE2D2</th>\n",
       "      <th>SAR1A</th>\n",
       "      <th>IER3</th>\n",
       "      <th>RASGEF1C</th>\n",
       "      <th>TTC7A</th>\n",
       "      <th>TPK1</th>\n",
       "      <th>PTGES3</th>\n",
       "      <th>HMG20B</th>\n",
       "      <th>GEM</th>\n",
       "      <th>...</th>\n",
       "      <th>STAB1</th>\n",
       "      <th>NFAT5</th>\n",
       "      <th>MAN1A2</th>\n",
       "      <th>EDA</th>\n",
       "      <th>SNCA</th>\n",
       "      <th>FAM135A</th>\n",
       "      <th>H2BC21</th>\n",
       "      <th>FCGR3A</th>\n",
       "      <th>ARL2-SNX15</th>\n",
       "      <th>VSIG4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EC3_AACCATGTCACGAAGG</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EC3_AATCGGTCACTTACGA</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EC3_ACAGCTAAGTGTCCCG</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EC3_ACAGCTACAAGGACAC</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EC3_ACAGCTAGTGATGTCT</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5567</th>\n",
       "      <td>EC10_TTTCCTCGTGGTAACG</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5568</th>\n",
       "      <td>EC10_TTTCCTCGTTTGACAC</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5569</th>\n",
       "      <td>EC10_TTTCCTCTCACAGGCC</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.693147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5570</th>\n",
       "      <td>EC10_TTTGTCACAAGCGTAG</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5571</th>\n",
       "      <td>EC10_TTTGTCATCTATCGCC</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.098612</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5572 rows × 1000 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "feature_name                  index    UBE2D2  SAR1A      IER3  RASGEF1C  \\\n",
       "0              EC3_AACCATGTCACGAAGG  0.000000    0.0  0.000000  0.000000   \n",
       "1              EC3_AATCGGTCACTTACGA  0.000000    0.0  0.000000  0.000000   \n",
       "2              EC3_ACAGCTAAGTGTCCCG  0.000000    0.0  0.000000  0.000000   \n",
       "3              EC3_ACAGCTACAAGGACAC  0.000000    0.0  0.000000  0.000000   \n",
       "4              EC3_ACAGCTAGTGATGTCT  0.693147    0.0  0.693147  0.000000   \n",
       "...                             ...       ...    ...       ...       ...   \n",
       "5567          EC10_TTTCCTCGTGGTAACG  0.000000    0.0  0.000000  0.000000   \n",
       "5568          EC10_TTTCCTCGTTTGACAC  0.000000    0.0  0.000000  0.000000   \n",
       "5569          EC10_TTTCCTCTCACAGGCC  0.693147    0.0  0.000000  0.693147   \n",
       "5570          EC10_TTTGTCACAAGCGTAG  0.000000    0.0  0.000000  0.000000   \n",
       "5571          EC10_TTTGTCATCTATCGCC  0.000000    0.0  0.000000  0.000000   \n",
       "\n",
       "feature_name  TTC7A      TPK1  PTGES3  HMG20B  GEM  ...     STAB1     NFAT5  \\\n",
       "0               0.0  0.000000     0.0     0.0  0.0  ...  0.000000  0.000000   \n",
       "1               0.0  0.000000     0.0     0.0  0.0  ...  0.000000  0.000000   \n",
       "2               0.0  0.000000     0.0     0.0  0.0  ...  0.693147  0.000000   \n",
       "3               0.0  0.000000     0.0     0.0  0.0  ...  0.000000  0.000000   \n",
       "4               0.0  0.000000     0.0     0.0  0.0  ...  0.000000  0.000000   \n",
       "...             ...       ...     ...     ...  ...  ...       ...       ...   \n",
       "5567            0.0  0.000000     0.0     0.0  0.0  ...  0.000000  0.000000   \n",
       "5568            0.0  0.000000     0.0     0.0  0.0  ...  0.693147  0.000000   \n",
       "5569            0.0  0.693147     0.0     0.0  0.0  ...  0.000000  0.693147   \n",
       "5570            0.0  0.000000     0.0     0.0  0.0  ...  0.000000  0.000000   \n",
       "5571            0.0  0.000000     0.0     0.0  0.0  ...  0.000000  0.000000   \n",
       "\n",
       "feature_name  MAN1A2  EDA  SNCA   FAM135A  H2BC21  FCGR3A  ARL2-SNX15  \\\n",
       "0                0.0  0.0   0.0  0.000000     0.0     0.0    0.000000   \n",
       "1                0.0  0.0   0.0  0.000000     0.0     0.0    0.000000   \n",
       "2                0.0  0.0   0.0  0.000000     0.0     0.0    0.000000   \n",
       "3                0.0  0.0   0.0  0.000000     0.0     0.0    0.000000   \n",
       "4                0.0  0.0   0.0  0.000000     0.0     0.0    0.000000   \n",
       "...              ...  ...   ...       ...     ...     ...         ...   \n",
       "5567             0.0  0.0   0.0  0.000000     0.0     0.0    0.000000   \n",
       "5568             0.0  0.0   0.0  0.693147     0.0     0.0    0.000000   \n",
       "5569             0.0  0.0   0.0  0.000000     0.0     0.0    0.693147   \n",
       "5570             0.0  0.0   0.0  0.693147     0.0     0.0    0.000000   \n",
       "5571             0.0  0.0   0.0  1.098612     0.0     0.0    0.000000   \n",
       "\n",
       "feature_name     VSIG4  \n",
       "0             0.000000  \n",
       "1             0.000000  \n",
       "2             0.000000  \n",
       "3             0.000000  \n",
       "4             0.000000  \n",
       "...                ...  \n",
       "5567          0.000000  \n",
       "5568          0.000000  \n",
       "5569          0.693147  \n",
       "5570          0.000000  \n",
       "5571          0.000000  \n",
       "\n",
       "[5572 rows x 1000 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "datExpr_donor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "93aa0f54",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge 'datExpr_reset' with 'metadata' on the 'index' and 'cell_id' columns\n",
    "datExpr_donor = pd.merge(datExpr_donor, metadata[['cell_id', 'donor_id']], left_on='index', right_on='cell_id', how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "edb12d51",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>UBE2D2</th>\n",
       "      <th>SAR1A</th>\n",
       "      <th>IER3</th>\n",
       "      <th>RASGEF1C</th>\n",
       "      <th>TTC7A</th>\n",
       "      <th>TPK1</th>\n",
       "      <th>PTGES3</th>\n",
       "      <th>HMG20B</th>\n",
       "      <th>GEM</th>\n",
       "      <th>...</th>\n",
       "      <th>MAN1A2</th>\n",
       "      <th>EDA</th>\n",
       "      <th>SNCA</th>\n",
       "      <th>FAM135A</th>\n",
       "      <th>H2BC21</th>\n",
       "      <th>FCGR3A</th>\n",
       "      <th>ARL2-SNX15</th>\n",
       "      <th>VSIG4</th>\n",
       "      <th>cell_id</th>\n",
       "      <th>donor_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>EC3_AACCATGTCACGAAGG</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>EC3_AACCATGTCACGAAGG</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>EC3_AATCGGTCACTTACGA</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>EC3_AATCGGTCACTTACGA</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>EC3_ACAGCTAAGTGTCCCG</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>EC3_ACAGCTAAGTGTCCCG</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EC3_ACAGCTACAAGGACAC</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>EC3_ACAGCTACAAGGACAC</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EC3_ACAGCTAGTGATGTCT</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>EC3_ACAGCTAGTGATGTCT</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5567</th>\n",
       "      <td>EC10_TTTCCTCGTGGTAACG</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>EC10_TTTCCTCGTGGTAACG</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5568</th>\n",
       "      <td>EC10_TTTCCTCGTTTGACAC</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>EC10_TTTCCTCGTTTGACAC</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5569</th>\n",
       "      <td>EC10_TTTCCTCTCACAGGCC</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>EC10_TTTCCTCTCACAGGCC</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5570</th>\n",
       "      <td>EC10_TTTGTCACAAGCGTAG</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>EC10_TTTGTCACAAGCGTAG</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5571</th>\n",
       "      <td>EC10_TTTGTCATCTATCGCC</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.098612</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>EC10_TTTGTCATCTATCGCC</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5572 rows × 1002 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      index    UBE2D2  SAR1A      IER3  RASGEF1C  TTC7A  \\\n",
       "0      EC3_AACCATGTCACGAAGG  0.000000    0.0  0.000000  0.000000    0.0   \n",
       "1      EC3_AATCGGTCACTTACGA  0.000000    0.0  0.000000  0.000000    0.0   \n",
       "2      EC3_ACAGCTAAGTGTCCCG  0.000000    0.0  0.000000  0.000000    0.0   \n",
       "3      EC3_ACAGCTACAAGGACAC  0.000000    0.0  0.000000  0.000000    0.0   \n",
       "4      EC3_ACAGCTAGTGATGTCT  0.693147    0.0  0.693147  0.000000    0.0   \n",
       "...                     ...       ...    ...       ...       ...    ...   \n",
       "5567  EC10_TTTCCTCGTGGTAACG  0.000000    0.0  0.000000  0.000000    0.0   \n",
       "5568  EC10_TTTCCTCGTTTGACAC  0.000000    0.0  0.000000  0.000000    0.0   \n",
       "5569  EC10_TTTCCTCTCACAGGCC  0.693147    0.0  0.000000  0.693147    0.0   \n",
       "5570  EC10_TTTGTCACAAGCGTAG  0.000000    0.0  0.000000  0.000000    0.0   \n",
       "5571  EC10_TTTGTCATCTATCGCC  0.000000    0.0  0.000000  0.000000    0.0   \n",
       "\n",
       "          TPK1  PTGES3  HMG20B  GEM  ...  MAN1A2  EDA  SNCA   FAM135A  H2BC21  \\\n",
       "0     0.000000     0.0     0.0  0.0  ...     0.0  0.0   0.0  0.000000     0.0   \n",
       "1     0.000000     0.0     0.0  0.0  ...     0.0  0.0   0.0  0.000000     0.0   \n",
       "2     0.000000     0.0     0.0  0.0  ...     0.0  0.0   0.0  0.000000     0.0   \n",
       "3     0.000000     0.0     0.0  0.0  ...     0.0  0.0   0.0  0.000000     0.0   \n",
       "4     0.000000     0.0     0.0  0.0  ...     0.0  0.0   0.0  0.000000     0.0   \n",
       "...        ...     ...     ...  ...  ...     ...  ...   ...       ...     ...   \n",
       "5567  0.000000     0.0     0.0  0.0  ...     0.0  0.0   0.0  0.000000     0.0   \n",
       "5568  0.000000     0.0     0.0  0.0  ...     0.0  0.0   0.0  0.693147     0.0   \n",
       "5569  0.693147     0.0     0.0  0.0  ...     0.0  0.0   0.0  0.000000     0.0   \n",
       "5570  0.000000     0.0     0.0  0.0  ...     0.0  0.0   0.0  0.693147     0.0   \n",
       "5571  0.000000     0.0     0.0  0.0  ...     0.0  0.0   0.0  1.098612     0.0   \n",
       "\n",
       "      FCGR3A  ARL2-SNX15     VSIG4                cell_id  donor_id  \n",
       "0        0.0    0.000000  0.000000   EC3_AACCATGTCACGAAGG         3  \n",
       "1        0.0    0.000000  0.000000   EC3_AATCGGTCACTTACGA         3  \n",
       "2        0.0    0.000000  0.000000   EC3_ACAGCTAAGTGTCCCG         3  \n",
       "3        0.0    0.000000  0.000000   EC3_ACAGCTACAAGGACAC         3  \n",
       "4        0.0    0.000000  0.000000   EC3_ACAGCTAGTGATGTCT         3  \n",
       "...      ...         ...       ...                    ...       ...  \n",
       "5567     0.0    0.000000  0.000000  EC10_TTTCCTCGTGGTAACG        10  \n",
       "5568     0.0    0.000000  0.000000  EC10_TTTCCTCGTTTGACAC        10  \n",
       "5569     0.0    0.693147  0.693147  EC10_TTTCCTCTCACAGGCC        10  \n",
       "5570     0.0    0.000000  0.000000  EC10_TTTGTCACAAGCGTAG        10  \n",
       "5571     0.0    0.000000  0.000000  EC10_TTTGTCATCTATCGCC        10  \n",
       "\n",
       "[5572 rows x 1002 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "datExpr_donor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "49c22377",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set the cell names as the index again\n",
    "datExpr_donor.set_index('index', inplace=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "77706931",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UBE2D2</th>\n",
       "      <th>SAR1A</th>\n",
       "      <th>IER3</th>\n",
       "      <th>RASGEF1C</th>\n",
       "      <th>TTC7A</th>\n",
       "      <th>TPK1</th>\n",
       "      <th>PTGES3</th>\n",
       "      <th>HMG20B</th>\n",
       "      <th>GEM</th>\n",
       "      <th>TREM1</th>\n",
       "      <th>...</th>\n",
       "      <th>MAN1A2</th>\n",
       "      <th>EDA</th>\n",
       "      <th>SNCA</th>\n",
       "      <th>FAM135A</th>\n",
       "      <th>H2BC21</th>\n",
       "      <th>FCGR3A</th>\n",
       "      <th>ARL2-SNX15</th>\n",
       "      <th>VSIG4</th>\n",
       "      <th>cell_id</th>\n",
       "      <th>donor_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>EC3_AACCATGTCACGAAGG</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>EC3_AACCATGTCACGAAGG</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC3_AATCGGTCACTTACGA</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>EC3_AATCGGTCACTTACGA</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC3_ACAGCTAAGTGTCCCG</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>EC3_ACAGCTAAGTGTCCCG</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC3_ACAGCTACAAGGACAC</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>EC3_ACAGCTACAAGGACAC</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC3_ACAGCTAGTGATGTCT</th>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>EC3_ACAGCTAGTGATGTCT</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC10_TTTCCTCGTGGTAACG</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>EC10_TTTCCTCGTGGTAACG</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC10_TTTCCTCGTTTGACAC</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>EC10_TTTCCTCGTTTGACAC</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC10_TTTCCTCTCACAGGCC</th>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>EC10_TTTCCTCTCACAGGCC</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC10_TTTGTCACAAGCGTAG</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>EC10_TTTGTCACAAGCGTAG</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC10_TTTGTCATCTATCGCC</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.098612</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>EC10_TTTGTCATCTATCGCC</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5572 rows × 1001 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         UBE2D2  SAR1A      IER3  RASGEF1C  TTC7A      TPK1  \\\n",
       "index                                                                         \n",
       "EC3_AACCATGTCACGAAGG   0.000000    0.0  0.000000  0.000000    0.0  0.000000   \n",
       "EC3_AATCGGTCACTTACGA   0.000000    0.0  0.000000  0.000000    0.0  0.000000   \n",
       "EC3_ACAGCTAAGTGTCCCG   0.000000    0.0  0.000000  0.000000    0.0  0.000000   \n",
       "EC3_ACAGCTACAAGGACAC   0.000000    0.0  0.000000  0.000000    0.0  0.000000   \n",
       "EC3_ACAGCTAGTGATGTCT   0.693147    0.0  0.693147  0.000000    0.0  0.000000   \n",
       "...                         ...    ...       ...       ...    ...       ...   \n",
       "EC10_TTTCCTCGTGGTAACG  0.000000    0.0  0.000000  0.000000    0.0  0.000000   \n",
       "EC10_TTTCCTCGTTTGACAC  0.000000    0.0  0.000000  0.000000    0.0  0.000000   \n",
       "EC10_TTTCCTCTCACAGGCC  0.693147    0.0  0.000000  0.693147    0.0  0.693147   \n",
       "EC10_TTTGTCACAAGCGTAG  0.000000    0.0  0.000000  0.000000    0.0  0.000000   \n",
       "EC10_TTTGTCATCTATCGCC  0.000000    0.0  0.000000  0.000000    0.0  0.000000   \n",
       "\n",
       "                       PTGES3  HMG20B  GEM  TREM1  ...  MAN1A2  EDA  SNCA  \\\n",
       "index                                              ...                      \n",
       "EC3_AACCATGTCACGAAGG      0.0     0.0  0.0    0.0  ...     0.0  0.0   0.0   \n",
       "EC3_AATCGGTCACTTACGA      0.0     0.0  0.0    0.0  ...     0.0  0.0   0.0   \n",
       "EC3_ACAGCTAAGTGTCCCG      0.0     0.0  0.0    0.0  ...     0.0  0.0   0.0   \n",
       "EC3_ACAGCTACAAGGACAC      0.0     0.0  0.0    0.0  ...     0.0  0.0   0.0   \n",
       "EC3_ACAGCTAGTGATGTCT      0.0     0.0  0.0    0.0  ...     0.0  0.0   0.0   \n",
       "...                       ...     ...  ...    ...  ...     ...  ...   ...   \n",
       "EC10_TTTCCTCGTGGTAACG     0.0     0.0  0.0    0.0  ...     0.0  0.0   0.0   \n",
       "EC10_TTTCCTCGTTTGACAC     0.0     0.0  0.0    0.0  ...     0.0  0.0   0.0   \n",
       "EC10_TTTCCTCTCACAGGCC     0.0     0.0  0.0    0.0  ...     0.0  0.0   0.0   \n",
       "EC10_TTTGTCACAAGCGTAG     0.0     0.0  0.0    0.0  ...     0.0  0.0   0.0   \n",
       "EC10_TTTGTCATCTATCGCC     0.0     0.0  0.0    0.0  ...     0.0  0.0   0.0   \n",
       "\n",
       "                        FAM135A  H2BC21  FCGR3A  ARL2-SNX15     VSIG4  \\\n",
       "index                                                                   \n",
       "EC3_AACCATGTCACGAAGG   0.000000     0.0     0.0    0.000000  0.000000   \n",
       "EC3_AATCGGTCACTTACGA   0.000000     0.0     0.0    0.000000  0.000000   \n",
       "EC3_ACAGCTAAGTGTCCCG   0.000000     0.0     0.0    0.000000  0.000000   \n",
       "EC3_ACAGCTACAAGGACAC   0.000000     0.0     0.0    0.000000  0.000000   \n",
       "EC3_ACAGCTAGTGATGTCT   0.000000     0.0     0.0    0.000000  0.000000   \n",
       "...                         ...     ...     ...         ...       ...   \n",
       "EC10_TTTCCTCGTGGTAACG  0.000000     0.0     0.0    0.000000  0.000000   \n",
       "EC10_TTTCCTCGTTTGACAC  0.693147     0.0     0.0    0.000000  0.000000   \n",
       "EC10_TTTCCTCTCACAGGCC  0.000000     0.0     0.0    0.693147  0.693147   \n",
       "EC10_TTTGTCACAAGCGTAG  0.693147     0.0     0.0    0.000000  0.000000   \n",
       "EC10_TTTGTCATCTATCGCC  1.098612     0.0     0.0    0.000000  0.000000   \n",
       "\n",
       "                                     cell_id  donor_id  \n",
       "index                                                   \n",
       "EC3_AACCATGTCACGAAGG    EC3_AACCATGTCACGAAGG         3  \n",
       "EC3_AATCGGTCACTTACGA    EC3_AATCGGTCACTTACGA         3  \n",
       "EC3_ACAGCTAAGTGTCCCG    EC3_ACAGCTAAGTGTCCCG         3  \n",
       "EC3_ACAGCTACAAGGACAC    EC3_ACAGCTACAAGGACAC         3  \n",
       "EC3_ACAGCTAGTGATGTCT    EC3_ACAGCTAGTGATGTCT         3  \n",
       "...                                      ...       ...  \n",
       "EC10_TTTCCTCGTGGTAACG  EC10_TTTCCTCGTGGTAACG        10  \n",
       "EC10_TTTCCTCGTTTGACAC  EC10_TTTCCTCGTTTGACAC        10  \n",
       "EC10_TTTCCTCTCACAGGCC  EC10_TTTCCTCTCACAGGCC        10  \n",
       "EC10_TTTGTCACAAGCGTAG  EC10_TTTGTCACAAGCGTAG        10  \n",
       "EC10_TTTGTCATCTATCGCC  EC10_TTTGTCATCTATCGCC        10  \n",
       "\n",
       "[5572 rows x 1001 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "datExpr_donor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "869327bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove the 'cell_id' column if needed\n",
    "datExpr_donor.drop(columns=['cell_id'], inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "60bc4e70",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UBE2D2</th>\n",
       "      <th>SAR1A</th>\n",
       "      <th>IER3</th>\n",
       "      <th>RASGEF1C</th>\n",
       "      <th>TTC7A</th>\n",
       "      <th>TPK1</th>\n",
       "      <th>PTGES3</th>\n",
       "      <th>HMG20B</th>\n",
       "      <th>GEM</th>\n",
       "      <th>TREM1</th>\n",
       "      <th>...</th>\n",
       "      <th>NFAT5</th>\n",
       "      <th>MAN1A2</th>\n",
       "      <th>EDA</th>\n",
       "      <th>SNCA</th>\n",
       "      <th>FAM135A</th>\n",
       "      <th>H2BC21</th>\n",
       "      <th>FCGR3A</th>\n",
       "      <th>ARL2-SNX15</th>\n",
       "      <th>VSIG4</th>\n",
       "      <th>donor_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>EC3_AACCATGTCACGAAGG</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC3_AATCGGTCACTTACGA</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC3_ACAGCTAAGTGTCCCG</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC3_ACAGCTACAAGGACAC</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC3_ACAGCTAGTGATGTCT</th>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC10_TTTCCTCGTGGTAACG</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC10_TTTCCTCGTTTGACAC</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC10_TTTCCTCTCACAGGCC</th>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC10_TTTGTCACAAGCGTAG</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EC10_TTTGTCATCTATCGCC</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.098612</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5572 rows × 1000 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         UBE2D2  SAR1A      IER3  RASGEF1C  TTC7A      TPK1  \\\n",
       "index                                                                         \n",
       "EC3_AACCATGTCACGAAGG   0.000000    0.0  0.000000  0.000000    0.0  0.000000   \n",
       "EC3_AATCGGTCACTTACGA   0.000000    0.0  0.000000  0.000000    0.0  0.000000   \n",
       "EC3_ACAGCTAAGTGTCCCG   0.000000    0.0  0.000000  0.000000    0.0  0.000000   \n",
       "EC3_ACAGCTACAAGGACAC   0.000000    0.0  0.000000  0.000000    0.0  0.000000   \n",
       "EC3_ACAGCTAGTGATGTCT   0.693147    0.0  0.693147  0.000000    0.0  0.000000   \n",
       "...                         ...    ...       ...       ...    ...       ...   \n",
       "EC10_TTTCCTCGTGGTAACG  0.000000    0.0  0.000000  0.000000    0.0  0.000000   \n",
       "EC10_TTTCCTCGTTTGACAC  0.000000    0.0  0.000000  0.000000    0.0  0.000000   \n",
       "EC10_TTTCCTCTCACAGGCC  0.693147    0.0  0.000000  0.693147    0.0  0.693147   \n",
       "EC10_TTTGTCACAAGCGTAG  0.000000    0.0  0.000000  0.000000    0.0  0.000000   \n",
       "EC10_TTTGTCATCTATCGCC  0.000000    0.0  0.000000  0.000000    0.0  0.000000   \n",
       "\n",
       "                       PTGES3  HMG20B  GEM  TREM1  ...     NFAT5  MAN1A2  EDA  \\\n",
       "index                                              ...                          \n",
       "EC3_AACCATGTCACGAAGG      0.0     0.0  0.0    0.0  ...  0.000000     0.0  0.0   \n",
       "EC3_AATCGGTCACTTACGA      0.0     0.0  0.0    0.0  ...  0.000000     0.0  0.0   \n",
       "EC3_ACAGCTAAGTGTCCCG      0.0     0.0  0.0    0.0  ...  0.000000     0.0  0.0   \n",
       "EC3_ACAGCTACAAGGACAC      0.0     0.0  0.0    0.0  ...  0.000000     0.0  0.0   \n",
       "EC3_ACAGCTAGTGATGTCT      0.0     0.0  0.0    0.0  ...  0.000000     0.0  0.0   \n",
       "...                       ...     ...  ...    ...  ...       ...     ...  ...   \n",
       "EC10_TTTCCTCGTGGTAACG     0.0     0.0  0.0    0.0  ...  0.000000     0.0  0.0   \n",
       "EC10_TTTCCTCGTTTGACAC     0.0     0.0  0.0    0.0  ...  0.000000     0.0  0.0   \n",
       "EC10_TTTCCTCTCACAGGCC     0.0     0.0  0.0    0.0  ...  0.693147     0.0  0.0   \n",
       "EC10_TTTGTCACAAGCGTAG     0.0     0.0  0.0    0.0  ...  0.000000     0.0  0.0   \n",
       "EC10_TTTGTCATCTATCGCC     0.0     0.0  0.0    0.0  ...  0.000000     0.0  0.0   \n",
       "\n",
       "                       SNCA   FAM135A  H2BC21  FCGR3A  ARL2-SNX15     VSIG4  \\\n",
       "index                                                                         \n",
       "EC3_AACCATGTCACGAAGG    0.0  0.000000     0.0     0.0    0.000000  0.000000   \n",
       "EC3_AATCGGTCACTTACGA    0.0  0.000000     0.0     0.0    0.000000  0.000000   \n",
       "EC3_ACAGCTAAGTGTCCCG    0.0  0.000000     0.0     0.0    0.000000  0.000000   \n",
       "EC3_ACAGCTACAAGGACAC    0.0  0.000000     0.0     0.0    0.000000  0.000000   \n",
       "EC3_ACAGCTAGTGATGTCT    0.0  0.000000     0.0     0.0    0.000000  0.000000   \n",
       "...                     ...       ...     ...     ...         ...       ...   \n",
       "EC10_TTTCCTCGTGGTAACG   0.0  0.000000     0.0     0.0    0.000000  0.000000   \n",
       "EC10_TTTCCTCGTTTGACAC   0.0  0.693147     0.0     0.0    0.000000  0.000000   \n",
       "EC10_TTTCCTCTCACAGGCC   0.0  0.000000     0.0     0.0    0.693147  0.693147   \n",
       "EC10_TTTGTCACAAGCGTAG   0.0  0.693147     0.0     0.0    0.000000  0.000000   \n",
       "EC10_TTTGTCATCTATCGCC   0.0  1.098612     0.0     0.0    0.000000  0.000000   \n",
       "\n",
       "                       donor_id  \n",
       "index                            \n",
       "EC3_AACCATGTCACGAAGG          3  \n",
       "EC3_AATCGGTCACTTACGA          3  \n",
       "EC3_ACAGCTAAGTGTCCCG          3  \n",
       "EC3_ACAGCTACAAGGACAC          3  \n",
       "EC3_ACAGCTAGTGATGTCT          3  \n",
       "...                         ...  \n",
       "EC10_TTTCCTCGTGGTAACG        10  \n",
       "EC10_TTTCCTCGTTTGACAC        10  \n",
       "EC10_TTTCCTCTCACAGGCC        10  \n",
       "EC10_TTTGTCACAAGCGTAG        10  \n",
       "EC10_TTTGTCATCTATCGCC        10  \n",
       "\n",
       "[5572 rows x 1000 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "datExpr_donor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "cb267cbb",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Save the expression matrix with donor_id\n",
    "datExpr_donor.to_csv('dataset/mic_datExpr_donorid_singlecell.csv', index = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "108df4a2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "437c0cb9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "efd64442",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Now that we have our gene expression dataframe, it is now possible to aggregate the data for pseudobulking."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "e552400c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Aggregate expression by donor ID (summing the values)\n",
    "pseudobulk_df = datExpr_donor.groupby('donor_id').sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "81a5ca25",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UBE2D2</th>\n",
       "      <th>SAR1A</th>\n",
       "      <th>IER3</th>\n",
       "      <th>RASGEF1C</th>\n",
       "      <th>TTC7A</th>\n",
       "      <th>TPK1</th>\n",
       "      <th>PTGES3</th>\n",
       "      <th>HMG20B</th>\n",
       "      <th>GEM</th>\n",
       "      <th>TREM1</th>\n",
       "      <th>...</th>\n",
       "      <th>STAB1</th>\n",
       "      <th>NFAT5</th>\n",
       "      <th>MAN1A2</th>\n",
       "      <th>EDA</th>\n",
       "      <th>SNCA</th>\n",
       "      <th>FAM135A</th>\n",
       "      <th>H2BC21</th>\n",
       "      <th>FCGR3A</th>\n",
       "      <th>ARL2-SNX15</th>\n",
       "      <th>VSIG4</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>donor_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>6.931472</td>\n",
       "      <td>3.465736</td>\n",
       "      <td>19.238222</td>\n",
       "      <td>29.347748</td>\n",
       "      <td>6.238325</td>\n",
       "      <td>4.564348</td>\n",
       "      <td>6.238325</td>\n",
       "      <td>1.386294</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>44.374969</td>\n",
       "      <td>11.495820</td>\n",
       "      <td>7.336937</td>\n",
       "      <td>5.950643</td>\n",
       "      <td>10.802673</td>\n",
       "      <td>4.852030</td>\n",
       "      <td>3.465736</td>\n",
       "      <td>4.158883</td>\n",
       "      <td>1.386294</td>\n",
       "      <td>18.139610</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>19.120440</td>\n",
       "      <td>1.386294</td>\n",
       "      <td>5.662961</td>\n",
       "      <td>65.928200</td>\n",
       "      <td>18.545074</td>\n",
       "      <td>19.525904</td>\n",
       "      <td>8.317766</td>\n",
       "      <td>6.931472</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>11.090355</td>\n",
       "      <td>11.495820</td>\n",
       "      <td>5.950643</td>\n",
       "      <td>11.090355</td>\n",
       "      <td>20.506733</td>\n",
       "      <td>11.495820</td>\n",
       "      <td>17.040997</td>\n",
       "      <td>5.545177</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>5.545177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2.079442</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.386294</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>3.465736</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>12.594433</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.079442</td>\n",
       "      <td>1.386294</td>\n",
       "      <td>1.386294</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.386294</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5.662961</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>6.238325</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.871201</td>\n",
       "      <td>4.852030</td>\n",
       "      <td>4.564348</td>\n",
       "      <td>2.079442</td>\n",
       "      <td>4.564348</td>\n",
       "      <td>2.079442</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>8.723231</td>\n",
       "      <td>5.545177</td>\n",
       "      <td>4.158883</td>\n",
       "      <td>2.772589</td>\n",
       "      <td>2.079442</td>\n",
       "      <td>6.238325</td>\n",
       "      <td>3.465736</td>\n",
       "      <td>4.852030</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.178054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>8.317766</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>3.465736</td>\n",
       "      <td>18.139610</td>\n",
       "      <td>28.184597</td>\n",
       "      <td>9.534162</td>\n",
       "      <td>9.416378</td>\n",
       "      <td>2.772589</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>13.169796</td>\n",
       "      <td>14.268409</td>\n",
       "      <td>7.336937</td>\n",
       "      <td>6.931472</td>\n",
       "      <td>12.882114</td>\n",
       "      <td>14.673874</td>\n",
       "      <td>6.238325</td>\n",
       "      <td>13.693045</td>\n",
       "      <td>3.465736</td>\n",
       "      <td>6.238325</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>28.079237</td>\n",
       "      <td>9.639523</td>\n",
       "      <td>13.117681</td>\n",
       "      <td>4.158883</td>\n",
       "      <td>30.158678</td>\n",
       "      <td>27.555988</td>\n",
       "      <td>24.090252</td>\n",
       "      <td>10.227308</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>73.301506</td>\n",
       "      <td>35.010708</td>\n",
       "      <td>23.920353</td>\n",
       "      <td>13.575262</td>\n",
       "      <td>33.624413</td>\n",
       "      <td>21.317663</td>\n",
       "      <td>6.931472</td>\n",
       "      <td>40.385986</td>\n",
       "      <td>8.317766</td>\n",
       "      <td>79.844391</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>70.819923</td>\n",
       "      <td>9.704061</td>\n",
       "      <td>35.180607</td>\n",
       "      <td>288.971863</td>\n",
       "      <td>161.505615</td>\n",
       "      <td>110.800446</td>\n",
       "      <td>85.388435</td>\n",
       "      <td>13.575262</td>\n",
       "      <td>1.791759</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>54.079033</td>\n",
       "      <td>159.883743</td>\n",
       "      <td>76.418350</td>\n",
       "      <td>73.423744</td>\n",
       "      <td>124.665649</td>\n",
       "      <td>142.947891</td>\n",
       "      <td>17.158781</td>\n",
       "      <td>64.018661</td>\n",
       "      <td>6.238325</td>\n",
       "      <td>67.236404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>74.925560</td>\n",
       "      <td>32.002552</td>\n",
       "      <td>33.506630</td>\n",
       "      <td>153.331543</td>\n",
       "      <td>115.193764</td>\n",
       "      <td>140.264847</td>\n",
       "      <td>50.547630</td>\n",
       "      <td>15.249238</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>56.262703</td>\n",
       "      <td>124.727928</td>\n",
       "      <td>53.032536</td>\n",
       "      <td>58.747612</td>\n",
       "      <td>86.826851</td>\n",
       "      <td>119.117081</td>\n",
       "      <td>119.641464</td>\n",
       "      <td>19.525904</td>\n",
       "      <td>39.627174</td>\n",
       "      <td>56.785954</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>116.632179</td>\n",
       "      <td>68.281776</td>\n",
       "      <td>40.032639</td>\n",
       "      <td>53.032536</td>\n",
       "      <td>29.923111</td>\n",
       "      <td>133.281265</td>\n",
       "      <td>182.492737</td>\n",
       "      <td>9.704061</td>\n",
       "      <td>6.356108</td>\n",
       "      <td>1.791759</td>\n",
       "      <td>...</td>\n",
       "      <td>97.629524</td>\n",
       "      <td>183.475540</td>\n",
       "      <td>52.796970</td>\n",
       "      <td>54.706512</td>\n",
       "      <td>91.391197</td>\n",
       "      <td>103.174698</td>\n",
       "      <td>48.180504</td>\n",
       "      <td>63.429741</td>\n",
       "      <td>432.745911</td>\n",
       "      <td>251.994003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>146.712830</td>\n",
       "      <td>17.328680</td>\n",
       "      <td>39.575058</td>\n",
       "      <td>142.371399</td>\n",
       "      <td>118.360527</td>\n",
       "      <td>141.037216</td>\n",
       "      <td>139.062012</td>\n",
       "      <td>18.427292</td>\n",
       "      <td>2.772589</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>123.094772</td>\n",
       "      <td>206.378754</td>\n",
       "      <td>58.813278</td>\n",
       "      <td>80.929451</td>\n",
       "      <td>161.925690</td>\n",
       "      <td>211.728058</td>\n",
       "      <td>32.120338</td>\n",
       "      <td>102.142883</td>\n",
       "      <td>53.373463</td>\n",
       "      <td>191.849930</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 999 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              UBE2D2      SAR1A       IER3    RASGEF1C       TTC7A  \\\n",
       "donor_id                                                             \n",
       "1           6.931472   3.465736  19.238222   29.347748    6.238325   \n",
       "2          19.120440   1.386294   5.662961   65.928200   18.545074   \n",
       "3           2.079442   0.000000   1.386294    0.693147    0.000000   \n",
       "4           6.238325   0.000000   3.871201    4.852030    4.564348   \n",
       "5           8.317766   0.693147   3.465736   18.139610   28.184597   \n",
       "6          28.079237   9.639523  13.117681    4.158883   30.158678   \n",
       "7          70.819923   9.704061  35.180607  288.971863  161.505615   \n",
       "8          74.925560  32.002552  33.506630  153.331543  115.193764   \n",
       "9         116.632179  68.281776  40.032639   53.032536   29.923111   \n",
       "10        146.712830  17.328680  39.575058  142.371399  118.360527   \n",
       "\n",
       "                TPK1      PTGES3     HMG20B       GEM     TREM1  ...  \\\n",
       "donor_id                                                         ...   \n",
       "1           4.564348    6.238325   1.386294  0.693147  0.000000  ...   \n",
       "2          19.525904    8.317766   6.931472  0.000000  0.000000  ...   \n",
       "3           0.693147    3.465736   0.693147  0.000000  0.000000  ...   \n",
       "4           2.079442    4.564348   2.079442  0.000000  0.000000  ...   \n",
       "5           9.534162    9.416378   2.772589  0.000000  0.000000  ...   \n",
       "6          27.555988   24.090252  10.227308  0.693147  0.000000  ...   \n",
       "7         110.800446   85.388435  13.575262  1.791759  0.000000  ...   \n",
       "8         140.264847   50.547630  15.249238  0.693147  0.000000  ...   \n",
       "9         133.281265  182.492737   9.704061  6.356108  1.791759  ...   \n",
       "10        141.037216  139.062012  18.427292  2.772589  0.000000  ...   \n",
       "\n",
       "               STAB1       NFAT5     MAN1A2        EDA        SNCA  \\\n",
       "donor_id                                                             \n",
       "1          44.374969   11.495820   7.336937   5.950643   10.802673   \n",
       "2          11.090355   11.495820   5.950643  11.090355   20.506733   \n",
       "3          12.594433    0.000000   2.079442   1.386294    1.386294   \n",
       "4           8.723231    5.545177   4.158883   2.772589    2.079442   \n",
       "5          13.169796   14.268409   7.336937   6.931472   12.882114   \n",
       "6          73.301506   35.010708  23.920353  13.575262   33.624413   \n",
       "7          54.079033  159.883743  76.418350  73.423744  124.665649   \n",
       "8          56.262703  124.727928  53.032536  58.747612   86.826851   \n",
       "9          97.629524  183.475540  52.796970  54.706512   91.391197   \n",
       "10        123.094772  206.378754  58.813278  80.929451  161.925690   \n",
       "\n",
       "             FAM135A      H2BC21      FCGR3A  ARL2-SNX15       VSIG4  \n",
       "donor_id                                                              \n",
       "1           4.852030    3.465736    4.158883    1.386294   18.139610  \n",
       "2          11.495820   17.040997    5.545177    0.693147    5.545177  \n",
       "3           0.693147    0.000000    1.386294    0.000000    5.662961  \n",
       "4           6.238325    3.465736    4.852030    0.000000    3.178054  \n",
       "5          14.673874    6.238325   13.693045    3.465736    6.238325  \n",
       "6          21.317663    6.931472   40.385986    8.317766   79.844391  \n",
       "7         142.947891   17.158781   64.018661    6.238325   67.236404  \n",
       "8         119.117081  119.641464   19.525904   39.627174   56.785954  \n",
       "9         103.174698   48.180504   63.429741  432.745911  251.994003  \n",
       "10        211.728058   32.120338  102.142883   53.373463  191.849930  \n",
       "\n",
       "[10 rows x 999 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pseudobulk_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "f6133097",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Save the pseudobulk expression matrix with donor_id\n",
    "pseudobulk_df.to_csv('dataset/mic_datExpr_pseudobulk.csv', index = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77ba2969",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c4366bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "#We now have the pseudobulked data and the corresponding metadata dataframe to start the correlation network analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b91d63e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f167f27",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
